diff --git a/beetsplug/lastgenre/__init__.py b/beetsplug/lastgenre/__init__.py
index b1ef9685ed..7ec6a4e3a9 100644
--- a/beetsplug/lastgenre/__init__.py
+++ b/beetsplug/lastgenre/__init__.py
@@ -394,13 +394,12 @@ def _get_genre(
 
         # Return with a combined or freshly fetched genre list.
         if new_genres:
-            suffix = "whitelist" if self.whitelist else "any"
-            label += f", {suffix}"
-            if keep_genres:
-                label = f"keep + {label}"
-
             resolved_genres = self._combine_genres(keep_genres, new_genres)
             if any(resolved_genres):
+                suffix = "whitelist" if self.whitelist else "any"
+                label += f", {suffix}"
+                if keep_genres:
+                    label = f"keep + {label}"
                 return self._to_delimited_genre_string(resolved_genres), label
 
         # Nothing found, leave original.
