diff --git a/conan/tools/cmake/cmakedeps2/cmakedeps.py b/conan/tools/cmake/cmakedeps2/cmakedeps.py
index 0abe04f6960..77fe95b1642 100644
--- a/conan/tools/cmake/cmakedeps2/cmakedeps.py
+++ b/conan/tools/cmake/cmakedeps2/cmakedeps.py
@@ -1,12 +1,7 @@
-import textwrap
-
-import jinja2
-from jinja2 import Template
-
-from conan.api.output import Color
 from conan.internal import check_duplicated_generator
 from conan.tools.cmake.cmakedeps2.config import ConfigTemplate2
 from conan.tools.cmake.cmakedeps2.config_version import ConfigVersionTemplate2
+from conan.tools.cmake.cmakedeps2.target_configuration import TargetConfigurationTemplate2
 from conan.tools.cmake.cmakedeps2.targets import TargetsTemplate2
 from conan.tools.files import save
 from conan.errors import ConanException
@@ -64,17 +59,12 @@ def _content(self):
             config_version = ConfigVersionTemplate2(self, dep)
             ret[config_version.filename] = config_version.content()
 
-            targets = TargetsTemplate2(self, require, dep)
+            targets = TargetsTemplate2(self, dep)
             ret[targets.filename] = targets.content()
-
+            target_configuration = TargetConfigurationTemplate2(self, dep)
+            ret[target_configuration.filename] = target_configuration.content()
         return ret
 
-    def _generate_files(self, require, dep, ret):
-        target_configuration = TargetConfigurationTemplate(self, require, dep, find_module_mode)
-        ret[target_configuration.filename] = target_configuration.render()
-
-
-
     def set_property(self, dep, prop, value, build_context=False):
         """
         Using this method you can overwrite the :ref:`property<CMakeDeps Properties>` values set by
diff --git a/conan/tools/cmake/cmakedeps2/config.py b/conan/tools/cmake/cmakedeps2/config.py
index eb36f08e28c..494868ed3c5 100644
--- a/conan/tools/cmake/cmakedeps2/config.py
+++ b/conan/tools/cmake/cmakedeps2/config.py
@@ -46,7 +46,11 @@ def _template(self):
         include(${CMAKE_CURRENT_LIST_DIR}/{{ targets_include_file }})
         include(CMakeFindDependencyMacro)
 
-        check_build_type_defined()
+        get_property(isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+        if(NOT isMultiConfig AND NOT CMAKE_BUILD_TYPE)
+           message(FATAL_ERROR "Please, set the CMAKE_BUILD_TYPE variable when calling to CMake "
+                               "adding the '-DCMAKE_BUILD_TYPE=<build_type>' argument.")
+        endif()
 
         foreach(_DEPENDENCY {{ pkg_var(pkg_name, 'FIND_DEPENDENCY_NAMES', '') }} )
             # Check that we have not already called a find_package with the transitive dependency
diff --git a/conan/tools/cmake/cmakedeps2/target_configuration.py b/conan/tools/cmake/cmakedeps2/target_configuration.py
index 4f5e52d8c47..3958a83060f 100644
--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -1,263 +1,78 @@
+import os
 import textwrap
 
+import jinja2
+from jinja2 import Template
+
 
 class TargetConfigurationTemplate2:
     """
     FooTarget-release.cmake
     """
+    def __init__(self, cmakedeps, conanfile):
+        self._cmakedeps = cmakedeps
+        self._conanfile = conanfile
+
+    def content(self):
+        t = Template(self._template, trim_blocks=True, lstrip_blocks=True,
+                     undefined=jinja2.StrictUndefined)
+        return t.render(self._context)
+
     @property
     def filename(self):
-        name = "" if not self.generating_module else "module-"
-        name += "{}-Target-{}.cmake".format(self.file_name, self.cmakedeps.configuration.lower())
-        return name
+        f = self._cmakedeps.get_cmake_filename(self._conanfile)
+        return f"{f}-Targets-{self._cmakedeps.configuration.lower()}.cmake"
 
     @property
-    def context(self):
-        deps_targets_names = self.get_deps_targets_names() \
-            if not self.require.build else []
-
-        components_targets_names = self.get_declared_components_targets_names()
-        components_names = [(components_target_name.replace("::", "_"), components_target_name)
-                            for components_target_name in components_targets_names]
-
-        is_win = self.conanfile.settings.get_safe("os") == "Windows"
-        auto_link = self.cmakedeps.get_property("cmake_set_interface_link_directories",
-                                                self.conanfile, check_type=bool)
-        if self.conanfile.cpp_info.exe:
-            exe_location = self.conanfile.cpp_info.location.replace("\\", "/")
-            pfolder_var_name = "{}_PACKAGE_FOLDER{}".format(self.pkg_name, self.config_suffix)
+    def _context(self):
+        exes = {}
+        cpp_info = self._conanfile.cpp_info
+        pkg_name = self._conanfile.ref.name
+        config = self._cmakedeps.configuration.upper()
+        package_folder = self._conanfile.package_folder.replace("\\", "/")
+        package_folder_var = f"{pkg_name}_PACKAGE_FOLDER_{config}"
+
+        if cpp_info.has_components:
+            for comp_name in self._conanfile.cpp_info.components:
+                if comp_name is not None:
+                    pass
         else:
-            exe_location = pfolder_var_name = ""
-        return {"pkg_name": self.pkg_name,
-                "root_target_name": self.root_target_name,
-                "config_suffix": self.config_suffix,
-                "config": self.configuration.upper(),
-                "deps_targets_names": ";".join(deps_targets_names),
-                "components_names": components_names,
-                "configuration": self.cmakedeps.configuration,
-                "set_interface_link_directories": auto_link and is_win,
-                "exe": self.conanfile.cpp_info.exe,
-                "exe_location": f"${{{pfolder_var_name}}}/{exe_location}"}
+            if cpp_info.exe:
+                target = cpp_info.get_property("cmake_target_name") or f"{pkg_name}::{cpp_info.exe}"
+                exe_location = self._path(cpp_info.location, package_folder, package_folder_var)
+                exes[target] = exe_location
+
+        return {"package_folder": package_folder,
+                "package_folder_var": package_folder_var,
+                "config": config,
+                "exes": exes}
+
+    @staticmethod
+    def _path(p, package_folder, package_folder_var):
+        def escape(p_):
+            return p_.replace("$", "\\$").replace('"', '\\"')
+
+        p = p.replace("\\", "/")
+        if os.path.isabs(p):
+            if p.startswith(package_folder):
+                rel = p[len(package_folder):]
+                return f"${{{package_folder_var}}}/{escape(rel)}"
+            return escape(p)
+
+        return f"${{{package_folder_var}}}/{escape(p)}"
 
     @property
-    def template(self):
+    def _template(self):
         return textwrap.dedent("""\
-        # Avoid multiple calls to find_package to append duplicated properties to the targets
-        include_guard()
+        set({{package_folder_var}} "{{package_folder}}")
 
-        {%- macro comp_var(pkg_name, comp_name, var, config_suffix) -%}
-            {{'${'+pkg_name+'_'+comp_name+'_'+var+config_suffix+'}'}}
-        {%- endmacro -%}
-        {%- macro pkg_var(pkg_name, var, config_suffix) -%}
-            {{'${'+pkg_name+'_'+var+config_suffix+'}'}}
-        {%- endmacro -%}
-
-        ########### VARIABLES #######################################################################
-        #############################################################################################
-        set({{ pkg_name }}_FRAMEWORKS_FOUND{{ config_suffix }} "") # Will be filled later
-        conan_find_apple_frameworks({{ pkg_name }}_FRAMEWORKS_FOUND{{ config_suffix }} "{{ pkg_var(pkg_name, 'FRAMEWORKS', config_suffix) }}" "{{ pkg_var(pkg_name, 'FRAMEWORK_DIRS', config_suffix) }}")
-
-        set({{ pkg_name }}_LIBRARIES_TARGETS "") # Will be filled later
-
-
-        ######## Create an interface target to contain all the dependencies (frameworks, system and conan deps)
-        if(NOT TARGET {{ pkg_name+'_DEPS_TARGET'}})
-            add_library({{ pkg_name+'_DEPS_TARGET'}} INTERFACE IMPORTED)
+        # Exe information
+        {% for exe, location in exes.items() %}
+        message(STATUS "Conan: Target declared imported executable '{{exe}}'")
+        if(NOT TARGET {{ exe }})
+            add_executable({{exe}} IMPORTED)
         endif()
-
-        set_property(TARGET {{ pkg_name + '_DEPS_TARGET'}}
-                     APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-                     $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'FRAMEWORKS_FOUND', config_suffix) }}>
-                     $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'SYSTEM_LIBS', config_suffix) }}>
-                     $<$<CONFIG:{{configuration}}>:{{ deps_targets_names }}>)
-
-        ####### Find the libraries declared in cpp_info.libs, create an IMPORTED target for each one and link the
-        ####### {{pkg_name}}_DEPS_TARGET to all of them
-        conan_package_library_targets("{{ pkg_var(pkg_name, 'LIBS', config_suffix) }}"    # libraries
-                                      "{{ pkg_var(pkg_name, 'LIB_DIRS', config_suffix) }}" # package_libdir
-                                      "{{ pkg_var(pkg_name, 'BIN_DIRS', config_suffix) }}" # package_bindir
-                                      "{{ pkg_var(pkg_name, 'LIBRARY_TYPE', config_suffix) }}"
-                                      "{{ pkg_var(pkg_name, 'IS_HOST_WINDOWS', config_suffix) }}"
-                                      {{ pkg_name + '_DEPS_TARGET'}}
-                                      {{ pkg_name }}_LIBRARIES_TARGETS  # out_libraries_targets
-                                      "{{ config_suffix }}"
-                                      "{{ pkg_name }}"    # package_name
-                                      "{{ pkg_var(pkg_name, 'NO_SONAME_MODE', config_suffix) }}")  # soname
-
-        # FIXME: What is the result of this for multi-config? All configs adding themselves to path?
-        set(CMAKE_MODULE_PATH {{ pkg_var(pkg_name, 'BUILD_DIRS', config_suffix) }} {{ '${' }}CMAKE_MODULE_PATH})
-        {% if not components_names %}
-
-        ########## GLOBAL TARGET PROPERTIES {{ configuration }} ########################################
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'OBJECTS', config_suffix) }}>
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'LIBRARIES_TARGETS', '') }}>
-                         )
-
-            if("{{ pkg_var(pkg_name, 'LIBS', config_suffix) }}" STREQUAL "")
-                # If the package is not declaring any "cpp_info.libs" the package deps, system libs,
-                # frameworks etc are not linked to the imported targets and we need to do it to the
-                # global target
-                set_property(TARGET {{root_target_name}}
-                             APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-                             {{pkg_name}}_DEPS_TARGET)
-            endif()
-
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_LINK_OPTIONS
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'LINKER_FLAGS', config_suffix) }}>)
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'INCLUDE_DIRS', config_suffix) }}>)
-            # Necessary to find LINK shared libraries in Linux
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_LINK_DIRECTORIES
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'LIB_DIRS', config_suffix) }}>)
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'COMPILE_DEFINITIONS', config_suffix) }}>)
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_COMPILE_OPTIONS
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'COMPILE_OPTIONS', config_suffix) }}>)
-
-            {%- if set_interface_link_directories %}
-
-            # This is only used for '#pragma comment(lib, "foo")' (automatic link)
-            set_property(TARGET {{root_target_name}}
-                         APPEND PROPERTY INTERFACE_LINK_DIRECTORIES
-                         $<$<CONFIG:{{configuration}}>:{{ pkg_var(pkg_name, 'LIB_DIRS', config_suffix) }}>)
-            {%- endif %}
-
-            # Exe information
-            {% if exe %}
-            message(STATUS "Conan: Target declared imported executable '{{pkg_name}}::{{exe}}'")
-            add_executable({{pkg_name}}::{{exe}} IMPORTED)
-            set_target_properties({{pkg_name}}::{{exe}} PROPERTIES IMPORTED_LOCATION${config_suffix} "{{exe_location}}")
-            {%- endif %}
-
-        {%- else %}
-
-        ########## COMPONENTS TARGET PROPERTIES {{ configuration }} ########################################
-
-            {%- for comp_variable_name, comp_target_name in components_names %}
-
-
-            ########## COMPONENT {{ comp_target_name }} #############
-
-                set({{ pkg_name }}_{{ comp_variable_name }}_FRAMEWORKS_FOUND{{ config_suffix }} "")
-                conan_find_apple_frameworks({{ pkg_name }}_{{ comp_variable_name }}_FRAMEWORKS_FOUND{{ config_suffix }} "{{ comp_var(pkg_name, comp_variable_name, 'FRAMEWORKS', config_suffix) }}" "{{ comp_var(pkg_name, comp_variable_name, 'FRAMEWORK_DIRS', config_suffix) }}")
-
-                set({{ pkg_name }}_{{ comp_variable_name }}_LIBRARIES_TARGETS "")
-
-                ######## Create an interface target to contain all the dependencies (frameworks, system and conan deps)
-                if(NOT TARGET {{ pkg_name + '_' + comp_variable_name + '_DEPS_TARGET'}})
-                    add_library({{ pkg_name + '_' + comp_variable_name + '_DEPS_TARGET'}} INTERFACE IMPORTED)
-                endif()
-
-                set_property(TARGET {{ pkg_name + '_' + comp_variable_name + '_DEPS_TARGET'}}
-                             APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-                             $<$<CONFIG:{{configuration}}>:{{ comp_var(pkg_name, comp_variable_name, 'FRAMEWORKS_FOUND', config_suffix) }}>
-                             $<$<CONFIG:{{configuration}}>:{{ comp_var(pkg_name, comp_variable_name, 'SYSTEM_LIBS', config_suffix) }}>
-                             $<$<CONFIG:{{configuration}}>:{{ comp_var(pkg_name, comp_variable_name, 'DEPENDENCIES', config_suffix) }}>
-                             )
-
-                ####### Find the libraries declared in cpp_info.component["xxx"].libs,
-                ####### create an IMPORTED target for each one and link the '{{pkg_name}}_{{comp_variable_name}}_DEPS_TARGET' to all of them
-                conan_package_library_targets("{{ comp_var(pkg_name, comp_variable_name, 'LIBS', config_suffix) }}"
-                                      "{{ comp_var(pkg_name, comp_variable_name, 'LIB_DIRS', config_suffix) }}"
-                                      "{{ comp_var(pkg_name, comp_variable_name, 'BIN_DIRS', config_suffix) }}" # package_bindir
-                                      "{{ comp_var(pkg_name, comp_variable_name, 'LIBRARY_TYPE', config_suffix) }}"
-                                      "{{ comp_var(pkg_name, comp_variable_name, 'IS_HOST_WINDOWS', config_suffix) }}"
-                                      {{ pkg_name + '_' + comp_variable_name + '_DEPS_TARGET'}}
-                                      {{ pkg_name + '_' + comp_variable_name + '_LIBRARIES_TARGETS'}}
-                                      "{{ config_suffix }}"
-                                      "{{ pkg_name }}_{{ comp_variable_name }}"
-                                      "{{ comp_var(pkg_name, comp_variable_name, 'NO_SONAME_MODE', config_suffix) }}")
-
-
-                ########## TARGET PROPERTIES #####################################
-                set_property(TARGET {{comp_target_name}}
-                             APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-                             $<$<CONFIG:{{configuration}}>:{{ comp_var(pkg_name, comp_variable_name, 'OBJECTS', config_suffix) }}>
-                             $<$<CONFIG:{{configuration}}>:{{ comp_var(pkg_name, comp_variable_name, 'LIBRARIES_TARGETS', '') }}>
-                             )
-
-                if("{{ comp_var(pkg_name, comp_variable_name, 'LIBS', config_suffix) }}" STREQUAL "")
-                    # If the component is not declaring any "cpp_info.components['foo'].libs" the system, frameworks etc are not
-                    # linked to the imported targets and we need to do it to the global target
-                    set_property(TARGET {{comp_target_name}}
-                                 APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-                                 {{pkg_name}}_{{comp_variable_name}}_DEPS_TARGET)
-                endif()
-
-                set_property(TARGET {{ comp_target_name }} APPEND PROPERTY INTERFACE_LINK_OPTIONS
-                             $<$<CONFIG:{{ configuration }}>:{{ comp_var(pkg_name, comp_variable_name, 'LINKER_FLAGS', config_suffix) }}>)
-                set_property(TARGET {{ comp_target_name }} APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES
-                             $<$<CONFIG:{{ configuration }}>:{{ comp_var(pkg_name, comp_variable_name, 'INCLUDE_DIRS', config_suffix) }}>)
-                set_property(TARGET {{comp_target_name }} APPEND PROPERTY INTERFACE_LINK_DIRECTORIES
-                             $<$<CONFIG:{{ configuration }}>:{{ comp_var(pkg_name, comp_variable_name, 'LIB_DIRS', config_suffix) }}>)
-                set_property(TARGET {{ comp_target_name }} APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS
-                             $<$<CONFIG:{{ configuration }}>:{{ comp_var(pkg_name, comp_variable_name, 'COMPILE_DEFINITIONS', config_suffix) }}>)
-                set_property(TARGET {{ comp_target_name }} APPEND PROPERTY INTERFACE_COMPILE_OPTIONS
-                             $<$<CONFIG:{{ configuration }}>:{{ comp_var(pkg_name, comp_variable_name, 'COMPILE_OPTIONS', config_suffix) }}>)
-
-                {%- if set_interface_link_directories %}
-                # This is only used for '#pragma comment(lib, "foo")' (automatic link)
-                set_property(TARGET {{ comp_target_name }} APPEND PROPERTY INTERFACE_LINK_DIRECTORIES
-                             $<$<CONFIG:{{ configuration }}>:{{ comp_var(pkg_name, comp_variable_name, 'LIB_DIRS', config_suffix) }}>)
-
-                {%- endif %}
-            {%endfor %}
-
-
-            ########## AGGREGATED GLOBAL TARGET WITH THE COMPONENTS #####################
-            {%- for comp_variable_name, comp_target_name in components_names %}
-
-            set_property(TARGET {{root_target_name}} APPEND PROPERTY INTERFACE_LINK_LIBRARIES {{ comp_target_name }})
-
-            {%- endfor %}
-
-
-        {%- endif %}
-
-
-        ########## For the modules (FindXXX)
-        set({{ pkg_name }}_LIBRARIES{{ config_suffix }} {{root_target_name}})
-
+        set_property(TARGET {{exe}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
+        set_target_properties({{exe}} PROPERTIES IMPORTED_LOCATION_{{config}} "{{location}}")
+        {%- endfor %}
         """)
-
-    def get_declared_components_targets_names(self):
-        """Returns a list of component_name"""
-        ret = []
-        sorted_comps = self.conanfile.cpp_info.get_sorted_components()
-        for comp_name, comp in sorted_comps.items():
-            ret.append(self.get_component_alias(self.conanfile, comp_name))
-        ret.reverse()
-        return ret
-
-    def get_deps_targets_names(self):
-        """
-          - [{foo}::{bar}, ] of the required
-        """
-        ret = []
-
-        # Get a list of dependencies target names
-        # Declared cppinfo.requires or .components[].requires
-        transitive_reqs = self.cmakedeps.get_transitive_requires(self.conanfile)
-        if self.conanfile.cpp_info.required_components:
-            for dep_name, component_name in self.conanfile.cpp_info.required_components:
-                try:
-                    # if not dep_name, it is internal, from current self.conanfile
-                    req = transitive_reqs[dep_name] if dep_name is not None else self.conanfile
-                except KeyError:
-                    # if it raises it means the required component is not in the direct_host
-                    # dependencies, maybe it has been filtered out by traits => Skip
-                    pass
-                else:
-                    component_name = self.get_component_alias(req, component_name)
-                    ret.append(component_name)
-        elif transitive_reqs:
-            # Regular external "conanfile.requires" declared, not cpp_info requires
-            ret = [self.get_root_target_name(r) for r in transitive_reqs.values()]
-        return ret
diff --git a/conan/tools/cmake/cmakedeps2/targets.py b/conan/tools/cmake/cmakedeps2/targets.py
index 19680e0cde6..94759c3675a 100644
--- a/conan/tools/cmake/cmakedeps2/targets.py
+++ b/conan/tools/cmake/cmakedeps2/targets.py
@@ -24,53 +24,20 @@ def filename(self):
 
     @property
     def _context(self):
-        data_pattern = "${CMAKE_CURRENT_LIST_DIR}/" if not self.generating_module else "${CMAKE_CURRENT_LIST_DIR}/module-"
-        data_pattern += "{}-*-data.cmake".format(self.file_name)
-
-        target_pattern = "" if not self.generating_module else "module-"
-        target_pattern += "{}-Target-*.cmake".format(self.file_name)
-
-        cmake_target_aliases = self.conanfile.cpp_info.\
-            get_property("cmake_target_aliases", check_type=list) or dict()
-
-        target = self.root_target_name
-        cmake_target_aliases = {alias: target for alias in cmake_target_aliases}
-
-        cmake_component_target_aliases = dict()
-        for comp_name in self.conanfile.cpp_info.components:
-            if comp_name is not None:
-                aliases = \
-                    self.conanfile.cpp_info.components[comp_name].\
-                    get_property("cmake_target_aliases", check_type=list) or dict()
-
-                target = self.get_component_alias(self.conanfile, comp_name)
-                cmake_component_target_aliases[comp_name] = {alias: target for alias in aliases}
-
-        ret = {"pkg_name": self.pkg_name,
-               "root_target_name": self.root_target_name,
-               "file_name": self.file_name,
-               "data_pattern": data_pattern,
-               "target_pattern": target_pattern,
-               "cmake_target_aliases": cmake_target_aliases,
-               "cmake_component_target_aliases": cmake_component_target_aliases}
-
+        filename = self._cmakedeps.get_cmake_filename(self._conanfile)
+        ret = {"ref": str(self._conanfile.ref),
+               "filename": filename}
         return ret
 
     @property
     def _template(self):
+        # TODO: What if target exists only in Debug or Release?
         return textwrap.dedent("""\
-        # Load the debug and release variables
-        file(GLOB DATA_FILES "{{data_pattern}}")
-
-        foreach(f ${DATA_FILES})
-            include(${f})
-        endforeach()
-
-
-        # Load the debug and release library finders
-        file(GLOB CONFIG_FILES "${CMAKE_CURRENT_LIST_DIR}/{{ target_pattern }}")
+            message(STATUS "Configuring Targets for {{ ref }}")
 
-        foreach(f ${CONFIG_FILES})
-            include(${f})
-        endforeach()
+            # Load information for each installed configuration.
+            file(GLOB _cmake_config_files "${CMAKE_CURRENT_LIST_DIR}/{{filename}}-Targets-*.cmake")
+            foreach(_cmake_config_file IN LISTS _cmake_config_files)
+              include("${_cmake_config_file}")
+            endforeach()
         """)
diff --git a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
index c8f9efb1da6..7d39f90bef3 100644
--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
@@ -57,7 +57,6 @@ class Consumer(ConanFile):
 
             def generate(self):
                 deps = CMakeDeps(self)
-                deps.build_context_activated = ["mytool"]
                 deps.generate()
                 tc = CMakeToolchain(self)
                 tc.generate()
@@ -80,7 +79,6 @@ def build(self):
     c.save({"conanfile.py": consumer,
             "CMakeLists.txt": cmake}, clean_first=True)
     c.run("build . -c tools.cmake.cmakedeps:new=True")
-    assert "Conan: Target declared 'mytool::mytool'" in c.out
     assert "Conan: Target declared imported executable 'mytool::myexe'" in c.out
     assert "Mytool generating out.c!!!!!" in c.out
 
