diff --git a/conan/tools/cmake/cmakedeps2/target_configuration.py b/conan/tools/cmake/cmakedeps2/target_configuration.py
index 3958a83060f..4a3c6b4cf32 100644
--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -25,22 +25,26 @@ def filename(self):
 
     @property
     def _context(self):
-        exes = {}
         cpp_info = self._conanfile.cpp_info
         pkg_name = self._conanfile.ref.name
         config = self._cmakedeps.configuration.upper()
         package_folder = self._conanfile.package_folder.replace("\\", "/")
         package_folder_var = f"{pkg_name}_PACKAGE_FOLDER_{config}"
 
+        exes = {}
+
+        def _add_exe(info):
+            if info.exe:
+                target = info.get_property("cmake_target_name") or f"{pkg_name}::{info.exe}"
+                exe_location = self._path(info.location, package_folder, package_folder_var)
+                exes[target] = exe_location
+
         if cpp_info.has_components:
-            for comp_name in self._conanfile.cpp_info.components:
-                if comp_name is not None:
-                    pass
+            for name, component in cpp_info.components.items():
+                _add_exe(component)
         else:
             if cpp_info.exe:
-                target = cpp_info.get_property("cmake_target_name") or f"{pkg_name}::{cpp_info.exe}"
-                exe_location = self._path(cpp_info.location, package_folder, package_folder_var)
-                exes[target] = exe_location
+                _add_exe(cpp_info)
 
         return {"package_folder": package_folder,
                 "package_folder_var": package_folder_var,
@@ -74,5 +78,6 @@ def _template(self):
         endif()
         set_property(TARGET {{exe}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
         set_target_properties({{exe}} PROPERTIES IMPORTED_LOCATION_{{config}} "{{location}}")
-        {%- endfor %}
+
+        {% endfor %}
         """)
diff --git a/conan/tools/cmake/cmakedeps2/targets.py b/conan/tools/cmake/cmakedeps2/targets.py
index 94759c3675a..8205ff43d6a 100644
--- a/conan/tools/cmake/cmakedeps2/targets.py
+++ b/conan/tools/cmake/cmakedeps2/targets.py
@@ -31,7 +31,6 @@ def _context(self):
 
     @property
     def _template(self):
-        # TODO: What if target exists only in Debug or Release?
         return textwrap.dedent("""\
             message(STATUS "Configuring Targets for {{ ref }}")
 
diff --git a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
index 7d39f90bef3..56f91f64d6e 100644
--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
@@ -30,7 +30,8 @@ def package(self):
                 cmake.install()
 
             def package_info(self):
-                self.cpp_info.exe = "myexe"
+                self.cpp_info.exe = "mytool"
+                self.cpp_info.set_property("cmake_target_name", "MyTool::myexe")
                 self.cpp_info.location = os.path.join("bin", "mytool")
         """)
     main = textwrap.dedent("""
@@ -73,13 +74,13 @@ def build(self):
         project(consumer C)
 
         find_package(mytool)
-        add_custom_command(OUTPUT out.c COMMAND mytool::myexe)
+        add_custom_command(OUTPUT out.c COMMAND MyTool::myexe)
         add_library(myLib out.c)
         """)
     c.save({"conanfile.py": consumer,
             "CMakeLists.txt": cmake}, clean_first=True)
     c.run("build . -c tools.cmake.cmakedeps:new=True")
-    assert "Conan: Target declared imported executable 'mytool::myexe'" in c.out
+    assert "Conan: Target declared imported executable 'MyTool::myexe'" in c.out
     assert "Mytool generating out.c!!!!!" in c.out
 
 
@@ -109,16 +110,18 @@ def package(self):
 
             def package_info(self):
                 self.cpp_info.components["my1exe"].exe = "mytool1"
-                self.cpp_info.components["my1exe"].location = os.path.join("bin", "mytool2")
+                self.cpp_info.components["my1exe"].set_property("cmake_target_name", "MyTool::my1exe")
+                self.cpp_info.components["my1exe"].location = os.path.join("bin", "mytool1")
                 self.cpp_info.components["my2exe"].exe = "mytool2"
-                self.cpp_info.components["my1exe"].location = os.path.join("bin", "mytool2")
+                self.cpp_info.components["my2exe"].set_property("cmake_target_name", "MyTool::my2exe")
+                self.cpp_info.components["my2exe"].location = os.path.join("bin", "mytool2")
         """)
     main = textwrap.dedent("""
         #include <iostream>
         #include <fstream>
 
         int main() {{
-            std::cout << "Mytool{number} generating out.c!!!!!" << std::endl;
+            std::cout << "Mytool{number} generating out{number}.c!!!!!" << std::endl;
             std::ofstream f("out{number}.c");
         }}
         """)
@@ -144,13 +147,13 @@ def package_info(self):
 
     consumer = textwrap.dedent("""
         from conan import ConanFile
-        from conan.tools.cmake import CMakeDeps2, CMakeToolchain, CMake, cmake_layout
+        from conan.tools.cmake import CMakeDeps, CMakeToolchain, CMake, cmake_layout
         class Consumer(ConanFile):
             settings = "os", "compiler", "arch", "build_type"
             tool_requires = "mytool/0.1"
 
             def generate(self):
-                deps = CMakeDeps2(self)
+                deps = CMakeDeps(self)
                 deps.generate()
                 tc = CMakeToolchain(self)
                 tc.generate()
@@ -168,15 +171,16 @@ def build(self):
 
         find_package(mytool)
         add_custom_command(OUTPUT out1.c
-                           COMMAND mytool::my1exe)
+                           COMMAND MyTool::my1exe)
         add_custom_command(OUTPUT out2.c
-                           COMMAND mytool::my2exe)
+                           COMMAND MyTool::my2exe)
         add_library(myLib out1.c out2.c)
         """)
     c.save({"conanfile.py": consumer,
             "CMakeLists.txt": cmake}, clean_first=True)
-    c.run("build .")
+    c.run("build . -c tools.cmake.cmakedeps:new=True")
     print(c.out)
-    assert "Conan: Target declared 'mytool::mytool'" in c.out
-    assert "Conan: Target declared imported executable 'mytool::myexe'" in c.out
-    assert "Mytool generating out.c!!!!!" in c.out
+    assert "Conan: Target declared imported executable 'MyTool::my1exe'" in c.out
+    assert "Mytool1 generating out1.c!!!!!" in c.out
+    assert "Conan: Target declared imported executable 'MyTool::my2exe'" in c.out
+    assert "Mytool2 generating out2.c!!!!!" in c.out
