diff --git a/sphinx/ext/autodoc/__init__.py b/sphinx/ext/autodoc/__init__.py
index 60c31e2542e..1a185c328e9 100644
--- a/sphinx/ext/autodoc/__init__.py
+++ b/sphinx/ext/autodoc/__init__.py
@@ -2202,16 +2202,17 @@ def format_args(self, **kwargs: Any) -> str:
                 #       But it makes users confused.
                 args = '()'
             else:
-                if inspect.isstaticmethod(self.object, cls=self.parent, name=self.object_name):
-                    self.env.events.emit(
-                        'autodoc-before-process-signature', self.object, False
-                    )
-                    sig = inspect.signature(self.object, bound_method=False,
-                                            type_aliases=self.config.autodoc_type_aliases)
-                else:
-                    self.env.events.emit('autodoc-before-process-signature', self.object, True)
-                    sig = inspect.signature(self.object, bound_method=True,
-                                            type_aliases=self.config.autodoc_type_aliases)
+                bound_method = not inspect.isstaticmethod(
+                    self.object, cls=self.parent, name=self.object_name,
+                )
+                self.env.events.emit(
+                    'autodoc-before-process-signature', self.object, bound_method,
+                )
+                sig = inspect.signature(
+                    self.object,
+                    bound_method=bound_method,
+                    type_aliases=self.config.autodoc_type_aliases,
+                )
                 args = stringify_signature(sig, **kwargs)
         except TypeError as exc:
             logger.warning(__("Failed to get a method signature for %s: %s"),
@@ -2234,8 +2235,10 @@ def add_directive_header(self, sig: str) -> None:
             self.add_line('   :abstractmethod:', sourcename)
         if inspect.iscoroutinefunction(obj) or inspect.isasyncgenfunction(obj):
             self.add_line('   :async:', sourcename)
-        if (inspect.isclassmethod(obj) or
-                inspect.is_singledispatch_method(obj) and inspect.isclassmethod(obj.func)):
+        if (
+            inspect.is_classmethod_like(obj)
+            or inspect.is_singledispatch_method(obj) and inspect.is_classmethod_like(obj.func)
+        ):
             self.add_line('   :classmethod:', sourcename)
         if inspect.isstaticmethod(obj, cls=self.parent, name=self.object_name):
             self.add_line('   :staticmethod:', sourcename)
diff --git a/sphinx/util/inspect.py b/sphinx/util/inspect.py
index e2bc8b1e6a7..b14eb88bddb 100644
--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -241,6 +241,38 @@ def isclassmethod(
     return False
 
 
+def isclassmethoddescriptor(
+    obj: Any,
+    cls: Any = None,
+    name: str | None = None,
+) -> TypeIs[types.ClassMethodDescriptorType]:
+    """Check if the object is a class method descriptor type.
+
+    This is typically useful to check if a built-in method is a class method.
+    """
+    if isinstance(obj, types.ClassMethodDescriptorType):
+        return True
+    if cls and name:
+        # trace __mro__ if the method is defined in parent class
+        sentinel = object()
+        for basecls in getmro(cls):
+            meth = basecls.__dict__.get(name, sentinel)
+            if meth is not sentinel:
+                return isinstance(obj, types.ClassMethodDescriptorType)
+    return False
+
+
+def is_classmethod_like(
+    obj: Any,
+    cls: Any = None,
+    name: str | None = None,
+) -> TypeIs[classmethod | types.ClassMethodDescriptorType]:
+    """Check if the object behaves like a class method."""
+    # Built-in methods are instances of ClassMethodDescriptorType
+    # while pure Python class methods are instances of classmethod().
+    return isclassmethod(obj, cls, name) or isclassmethoddescriptor(obj, cls, name)
+
+
 def isstaticmethod(
     obj: Any,
     cls: Any = None,
@@ -249,6 +281,8 @@ def isstaticmethod(
     """Check if the object is a :class:`staticmethod`."""
     if isinstance(obj, staticmethod):
         return True
+    # Unlike built-in class methods, built-in static methods
+    # satisfy "isinstance(cls.__dict__[name], staticmethod)".
     if cls and name:
         # trace __mro__ if the method is defined in parent class
         sentinel = object()
