diff --git a/sphinx/ext/autodoc/__init__.py b/sphinx/ext/autodoc/__init__.py
index de084bd6c70..0b6e044cb58 100644
--- a/sphinx/ext/autodoc/__init__.py
+++ b/sphinx/ext/autodoc/__init__.py
@@ -2385,17 +2385,26 @@ def format_args(self, **kwargs: Any) -> str:
                 #       But it makes users confused.
                 args = '()'
             else:
-                bound_method = not inspect.isstaticmethod(
-                    self.object, cls=self.parent, name=self.object_name,
-                )
-                self.env.events.emit(
-                    'autodoc-before-process-signature', self.object, bound_method,
-                )
-                sig = inspect.signature(
-                    self.object,
-                    bound_method=bound_method,
-                    type_aliases=self.config.autodoc_type_aliases,
-                )
+                if inspect.isstaticmethod(
+                    self.object, cls=self.parent, name=self.object_name
+                ):
+                    self.env.events.emit(
+                        'autodoc-before-process-signature', self.object, False
+                    )
+                    sig = inspect.signature(
+                        self.object,
+                        bound_method=False,
+                        type_aliases=self.config.autodoc_type_aliases,
+                    )
+                else:
+                    self.env.events.emit(
+                        'autodoc-before-process-signature', self.object, True
+                    )
+                    sig = inspect.signature(
+                        self.object,
+                        bound_method=True,
+                        type_aliases=self.config.autodoc_type_aliases,
+                    )
                 args = stringify_signature(sig, **kwargs)
         except TypeError as exc:
             logger.warning(
@@ -2420,9 +2429,9 @@ def add_directive_header(self, sig: str) -> None:
         if inspect.iscoroutinefunction(obj) or inspect.isasyncgenfunction(obj):
             self.add_line('   :async:', sourcename)
         if (
-            inspect.is_classmethod_like(obj)
+            inspect.is_class_method_like(obj)
             or inspect.is_singledispatch_method(obj)
-            and inspect.is_classmethod_like(obj.func)
+            and inspect.is_class_method_like(obj.func)
         ):
             self.add_line('   :classmethod:', sourcename)
         if inspect.isstaticmethod(obj, cls=self.parent, name=self.object_name):
diff --git a/sphinx/util/inspect.py b/sphinx/util/inspect.py
index 676ee09ab99..6ef62cb9359 100644
--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -222,9 +222,7 @@ def ispartial(obj: Any) -> TypeIs[partial | partialmethod]:
 
 
 def isclassmethod(
-    obj: Any,
-    cls: Any = None,
-    name: str | None = None,
+    obj: Any, cls: Any = None, name: str | None = None
 ) -> TypeIs[classmethod]:
     """Check if the object is a :class:`classmethod`."""
     if isinstance(obj, classmethod):
@@ -241,10 +239,8 @@ def isclassmethod(
     return False
 
 
-def isclassmethoddescriptor(
-    obj: Any,
-    cls: Any = None,
-    name: str | None = None,
+def is_class_method_descriptor(
+    obj: Any, cls: Any = None, name: str | None = None
 ) -> TypeIs[types.ClassMethodDescriptorType]:
     """Check if the object is a class method descriptor type.
 
@@ -262,21 +258,17 @@ def isclassmethoddescriptor(
     return False
 
 
-def is_classmethod_like(
-    obj: Any,
-    cls: Any = None,
-    name: str | None = None,
+def is_class_method_like(
+    obj: Any, cls: Any = None, name: str | None = None
 ) -> TypeIs[classmethod | types.ClassMethodDescriptorType]:
     """Check if the object behaves like a class method."""
     # Built-in methods are instances of ClassMethodDescriptorType
     # while pure Python class methods are instances of classmethod().
-    return isclassmethod(obj, cls, name) or isclassmethoddescriptor(obj, cls, name)
+    return isclassmethod(obj, cls, name) or is_class_method_descriptor(obj, cls, name)
 
 
 def isstaticmethod(
-    obj: Any,
-    cls: Any = None,
-    name: str | None = None,
+    obj: Any, cls: Any = None, name: str | None = None
 ) -> TypeIs[staticmethod]:
     """Check if the object is a :class:`staticmethod`."""
     if isinstance(obj, staticmethod):
