diff --git a/sphinx/util/inspect.py b/sphinx/util/inspect.py
index 72322bf9c32..3f396535dde 100644
--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -258,7 +258,9 @@ def is_classmethod_descriptor(
     return False
 
 
-def is_native_classmethod(obj: Any, cls: Any = None, name: str | None = None):
+def is_builtin_classmethod(
+    obj: Any, cls: Any = None, name: str | None = None
+) -> bool:
     """Check if the object is a class method implemented in C."""
     if is_classmethod_descriptor(obj, cls, name):
         return True
@@ -274,16 +276,14 @@ def is_native_classmethod(obj: Any, cls: Any = None, name: str | None = None):
         for basecls in getmro(cls):
             meth = basecls.__dict__.get(name, sentinel)
             if meth is not sentinel:
-                return is_native_classmethod(obj)
+                return is_builtin_classmethod(obj)
     return False
 
 def is_classmethod_like(
     obj: Any, cls: Any = None, name: str | None = None
-) -> TypeIs[classmethod | types.ClassMethodDescriptorType]:
+) -> bool:
     """Check if the object behaves like a class method."""
-    # Built-in methods are instances of ClassMethodDescriptorType
-    # while pure Python class methods are instances of classmethod().
-    return isclassmethod(obj, cls, name) or is_native_classmethod(obj, cls, name)
+    return isclassmethod(obj, cls, name) or is_builtin_classmethod(obj, cls, name)
 
 
 def isstaticmethod(
