diff --git a/conan/tools/cmake/cmakedeps2/target_configuration.py b/conan/tools/cmake/cmakedeps2/target_configuration.py
index b26af09e4ca..2c16643cab1 100644
--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -4,6 +4,7 @@
 import jinja2
 from jinja2 import Template
 
+from conan.errors import ConanException
 from conans.model.pkg_type import PackageType
 
 
@@ -18,6 +19,7 @@ def __init__(self, cmakedeps, conanfile):
     def content(self):
         t = Template(self._template, trim_blocks=True, lstrip_blocks=True,
                      undefined=jinja2.StrictUndefined)
+        print(t.render(self._context))
         return t.render(self._context)
 
     @property
@@ -25,6 +27,33 @@ def filename(self):
         f = self._cmakedeps.get_cmake_filename(self._conanfile)
         return f"{f}-Targets-{self._cmakedeps.configuration.lower()}.cmake"
 
+    def _requires(self, info, components):
+        result = []
+        requires = info.parsed_requires()
+        pkg_name = self._conanfile.ref.name
+        if not requires and not components:
+            # else depend on the components of the dependencies (default-components or all)
+            pass
+        for required_scope, required_comp in requires:
+            if required_scope is None:  # Points to a component of same package
+                dep_component = components[required_comp]
+                dep_target = dep_component.get_property("cmake_target_name") or f"{pkg_name}::{required_comp}"
+                result.append(dep_target)
+            else:  # Different package
+                try:  # Make sure the declared dependency is at least in the recipe requires
+                    self._conanfile.dependencies[required_pkg]
+                except KeyError:
+                    raise ConanException(f"{self.conanfile}: component '{comp_name}' required "
+                                         f"'{required_pkg}::{required_comp}', "
+                                         f"but '{required_pkg}' is not a direct dependency")
+                try:
+                    req = transitive_requires[required_pkg]
+                except KeyError:  # The transitive dep might have been skipped
+                    pass
+                else:
+                    public_comp_deps.append(self.get_component_alias(req, required_comp))
+        return result
+
     @property
     def _context(self):
         cpp_info = self._conanfile.cpp_info
@@ -34,26 +63,32 @@ def _context(self):
         package_folder_var = f"{pkg_name}_PACKAGE_FOLDER_{config}"
 
         libs = {}
-        if cpp_info.has_components:
-            for name, component in cpp_info.components.items():
-                pass
-        else:
-            if cpp_info.libs:
-                assert len(cpp_info.libs) == 1, "New CMakeDeps only allows 1 lib per component"
-                lib_name = cpp_info.libs[0]
-                target_name = cpp_info.get_property("cmake_target_name") or f"{pkg_name}::{lib_name}"
-                cpp_info.deduce_cps(self._conanfile.package_type)
-                location = self._path(cpp_info.location, package_folder, package_folder_var)
-                lib_type = "SHARED" if cpp_info.type is PackageType.SHARED else \
-                    "STATIC" if cpp_info.type is PackageType.STATIC else \
-                    "INTERFACE" if cpp_info.type is PackageType.HEADER else None
+
+        def _add_libs(info):
+            if info.libs:
+                assert len(info.libs) == 1, "New CMakeDeps only allows 1 lib per component"
+                lib_name = info.libs[0]
+                target_name = info.get_property("cmake_target_name") or f"{pkg_name}::{lib_name}"
+                info.deduce_cps(self._conanfile.package_type)
+                location = self._path(info.location, package_folder, package_folder_var)
+                lib_type = "SHARED" if info.type is PackageType.SHARED else \
+                    "STATIC" if info.type is PackageType.STATIC else \
+                    "INTERFACE" if info.type is PackageType.HEADER else None
                 includedirs = ";".join(self._path(i, package_folder, package_folder_var)
-                                       for i in cpp_info.includedirs) if cpp_info.includedirs else ""
+                                       for i in info.includedirs) if info.includedirs else ""
+                requires = ";".join(self._requires(info, self._conanfile.cpp_info.components))
                 if lib_type:
                     libs[target_name] = {"type": lib_type,
                                          "includedirs": includedirs,
                                          "location": location,
-                                         "link_location": ""}
+                                         "link_location": "",
+                                         "requires": requires}
+
+        if cpp_info.has_components:
+            for name, component in cpp_info.components.items():
+                _add_libs(component)
+        else:
+            _add_libs(cpp_info)
 
         exes = {}
 
@@ -84,7 +119,7 @@ def escape(p_):
         p = p.replace("\\", "/")
         if os.path.isabs(p):
             if p.startswith(package_folder):
-                rel = p[len(package_folder):]
+                rel = p[len(package_folder):].lstrip("/")
                 return f"${{{package_folder_var}}}/{escape(rel)}"
             return escape(p)
         return f"${{{package_folder_var}}}/{escape(p)}"
@@ -113,6 +148,11 @@ def _template(self):
         set_target_properties({{lib}} PROPERTIES IMPORTED_IMPLIB_{{config}}
                               "{{lib_info["link_location"]}}")
         {% endif %}
+        {% if lib_info.get("requires") %}
+        set_target_properties({{lib}} PROPERTIES INTERFACE_LINK_LIBRARIES
+                              $<$<CONFIG:{{config}}>:{{lib_info["requires"]}}>)
+        {% endif %}
+
         {% endfor %}
 
         # Exe information
diff --git a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
similarity index 87%
rename from test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
rename to test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
index 4feca4446cd..3d96b8a053d 100644
--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_exes.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
@@ -181,12 +181,32 @@ def build(self):
         assert "Mytool2 generating out2.c!!!!!" in c.out
 
 
+@pytest.mark.tool("cmake")
 class TestLibs:
-    def test_libs(self):
-        c = TestClient()
-        c.run("new cmake_lib -d name=pkg -d version=0.1")
-        c.run("create . -tf=")
-        c.save({}, clean_first=True)
-        c.run("new cmake_lib -d name=app -d version=0.1 -d requires=pkg/0.1")
+    def test_libs(self, matrix_client):
+        c = matrix_client
+        c.run("new cmake_lib -d name=app -d version=0.1 -d requires=matrix/1.0")
+        c.run("build . -c tools.cmake.cmakedeps:new=True")
+        assert "Conan: Target declared imported STATIC library 'matrix::matrix'" in c.out
+
+    def test_libs_components(self, matrix_client_components):
+        c = matrix_client_components
+
+        c.run("new cmake_exe -d name=app -d version=0.1 -d requires=matrix/1.0")
+        cmake = textwrap.dedent("""
+            cmake_minimum_required(VERSION 3.15)
+            project(app CXX)
+
+            find_package(matrix CONFIG REQUIRED)
+            add_executable(app src/app.cpp)
+            target_link_libraries(app PRIVATE matrix::module)
+            """)
+        app_cpp = textwrap.dedent("""
+            #include "module.h"
+            int main() { module();}
+            """)
+        c.save({"CMakelists.txt": cmake,
+                "src/app.cpp": app_cpp})
         c.run("build . -c tools.cmake.cmakedeps:new=True")
-        assert "Conan: Target declared imported STATIC library 'pkg::pkg'" in c.out
+        assert "Conan: Target declared imported STATIC library 'matrix::vector'" in c.out
+        assert "Conan: Target declared imported STATIC library 'matrix::module'" in c.out
diff --git a/test/functional/toolchains/conftest.py b/test/functional/toolchains/conftest.py
index e1f294b62a6..195e6d00e53 100644
--- a/test/functional/toolchains/conftest.py
+++ b/test/functional/toolchains/conftest.py
@@ -1,8 +1,10 @@
 import os
 import shutil
+import textwrap
 
 import pytest
 
+from conan.test.assets.sources import gen_function_h, gen_function_cpp
 from conan.test.utils.test_files import temp_folder
 from conan.test.utils.tools import TestClient
 
@@ -48,3 +50,77 @@ def transitive_libraries(_transitive_libraries):
     c.cache_folder = os.path.join(temp_folder(), ".conan2")
     shutil.copytree(_transitive_libraries.cache_folder, c.cache_folder)
     return c
+
+
+@pytest.fixture(scope="session")
+def _matrix_client_components():
+    """
+    2 components, different than the package name
+    """
+    c = TestClient()
+    vector_h = gen_function_h(name="vector")
+    vector_cpp = gen_function_cpp(name="vector", includes=["vector"])
+    module_h = gen_function_h(name="module")
+    module_cpp = gen_function_cpp(name="module", includes=["module", "vector"], calls=["vector"])
+
+    conanfile = textwrap.dedent("""
+      from os.path import join
+      from conan import ConanFile
+      from conan.tools.cmake import CMake
+      from conan.tools.files import copy
+
+      class Matrix(ConanFile):
+          name = "matrix"
+          version = "1.0"
+          settings = "os", "compiler", "build_type", "arch"
+          generators = "CMakeToolchain"
+          exports_sources = "src/*", "CMakeLists.txt"
+
+          def build(self):
+              cmake = CMake(self)
+              cmake.configure()
+              cmake.build()
+
+          def package(self):
+              cmake = CMake(self)
+              cmake.install()
+
+          def package_info(self):
+              self.cpp_info.components["vector"].libs = ["vector"]
+              self.cpp_info.components["vector"].includedirs = ["include"]
+              self.cpp_info.components["vector"].libdirs = ["lib"]
+              self.cpp_info.components["module"].libs = ["module"]
+              self.cpp_info.components["module"].includedirs = ["include"]
+              self.cpp_info.components["module"].libdirs = ["lib"]
+              self.cpp_info.components["module"].requires = ["vector"]
+          """)
+
+    cmakelists = textwrap.dedent("""
+       set(CMAKE_CXX_COMPILER_WORKS 1)
+       set(CMAKE_CXX_ABI_COMPILED 1)
+       cmake_minimum_required(VERSION 3.15)
+       project(matrix CXX)
+
+       add_library(vector src/vector.cpp)
+       add_library(module src/module.cpp)
+       target_link_libraries(module PRIVATE vector)
+
+       set_target_properties(module PROPERTIES PUBLIC_HEADER "src/module.h")
+       install(TARGETS vector module)
+       """)
+    c.save({"src/vector.h": vector_h,
+            "src/vector.cpp": vector_cpp,
+            "src/module.h": module_h,
+            "src/module.cpp": module_cpp,
+            "CMakeLists.txt": cmakelists,
+            "conanfile.py": conanfile})
+    c.run("create .")
+    return c
+
+
+@pytest.fixture()
+def matrix_client_components(_matrix_client_components):
+    c = TestClient()
+    c.cache_folder = os.path.join(temp_folder(), ".conan2")
+    shutil.copytree(_matrix_client_components.cache_folder, c.cache_folder)
+    return c
