diff --git a/conan/tools/cmake/cmakedeps2/cmakedeps.py b/conan/tools/cmake/cmakedeps2/cmakedeps.py
index 77fe95b1642..cb2b54dc901 100644
--- a/conan/tools/cmake/cmakedeps2/cmakedeps.py
+++ b/conan/tools/cmake/cmakedeps2/cmakedeps.py
@@ -103,7 +103,7 @@ def get_cmake_filename(self, dep, module_mode=None):
         # - The filename to generate (XXX-config.cmake or FindXXX.cmake)
         # - The name of the defined XXX_DIR variables
         # - The name of transitive dependencies for calls to find_dependency
-        if module_mode and self.get_find_mode(dep) in [FIND_MODE_MODULE, FIND_MODE_BOTH]:
+        if module_mode and self._get_find_mode(dep) in [FIND_MODE_MODULE, FIND_MODE_BOTH]:
             ret = self.get_property("cmake_module_file_name", dep)
             if ret:
                 return ret
@@ -114,7 +114,7 @@ def get_cmake_filename(self, dep, module_mode=None):
     def config_suffix(self):
         return "_{}".format(self.configuration.upper()) if self.configuration else ""
 
-    def get_find_mode(self, dep):
+    def _get_find_mode(self, dep):
         """
         :param dep: requirement
         :return: "none" or "config" or "module" or "both" or "config" when not set
diff --git a/conan/tools/cmake/cmakedeps2/target_configuration.py b/conan/tools/cmake/cmakedeps2/target_configuration.py
index 2c16643cab1..770cf38b6ce 100644
--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -14,7 +14,7 @@ class TargetConfigurationTemplate2:
     """
     def __init__(self, cmakedeps, conanfile):
         self._cmakedeps = cmakedeps
-        self._conanfile = conanfile
+        self._conanfile = conanfile  # The dependency conanfile, not the consumer one
 
     def content(self):
         t = Template(self._template, trim_blocks=True, lstrip_blocks=True,
@@ -31,9 +31,19 @@ def _requires(self, info, components):
         result = []
         requires = info.parsed_requires()
         pkg_name = self._conanfile.ref.name
-        if not requires and not components:
-            # else depend on the components of the dependencies (default-components or all)
-            pass
+        transitive_reqs = self._cmakedeps.get_transitive_requires(self._conanfile)
+        if not requires and not components:  # global cpp_info without components definition
+            # TODO: Not tested at all
+            for dependency in transitive_reqs.values():
+                dep_pkg_name = dependency.ref.name
+                dep_cpp_info = dependency.cpp_info
+                if dep_cpp_info.has_components:
+                    for name, component in dep_cpp_info.components.items():
+                        result.append(f"{dep_pkg_name}::{name}")
+                else:
+                    result.append("{dep_pkg_name}::{dep_pkg_name}")
+            return result
+
         for required_scope, required_comp in requires:
             if required_scope is None:  # Points to a component of same package
                 dep_component = components[required_comp]
@@ -76,7 +86,7 @@ def _add_libs(info):
                     "INTERFACE" if info.type is PackageType.HEADER else None
                 includedirs = ";".join(self._path(i, package_folder, package_folder_var)
                                        for i in info.includedirs) if info.includedirs else ""
-                requires = ";".join(self._requires(info, self._conanfile.cpp_info.components))
+                requires = " ".join(self._requires(info, self._conanfile.cpp_info.components))
                 if lib_type:
                     libs[target_name] = {"type": lib_type,
                                          "includedirs": includedirs,
@@ -90,6 +100,12 @@ def _add_libs(info):
         else:
             _add_libs(cpp_info)
 
+        if libs and f"{pkg_name}::{pkg_name}" not in libs:
+            # Add a generic interface target for the package depending on the others
+            all_requires = " ".join(libs.keys())
+            libs[f"{pkg_name}::{pkg_name}"] = {"type": "INTERFACE",
+                                               "requires": all_requires}
+
         exes = {}
 
         def _add_exe(info):
@@ -126,6 +142,7 @@ def escape(p_):
 
     @property
     def _template(self):
+        # TODO: Check why not set_property instead of target_link_libraries
         return textwrap.dedent("""\
         set({{package_folder_var}} "{{package_folder}}")
 
@@ -139,8 +156,8 @@ def _template(self):
         set_property(TARGET {{lib}} APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES
                      $<$<CONFIG:{{config}}>:{{lib_info["includedirs"]}}>)
         {% endif %}
-        set_property(TARGET {{lib}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
         {% if lib_info["type"] != "INTERFACE" %}
+        set_property(TARGET {{lib}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
         set_target_properties({{lib}} PROPERTIES IMPORTED_LOCATION_{{config}}
                               "{{lib_info["location"]}}")
         {% endif %}
@@ -149,8 +166,7 @@ def _template(self):
                               "{{lib_info["link_location"]}}")
         {% endif %}
         {% if lib_info.get("requires") %}
-        set_target_properties({{lib}} PROPERTIES INTERFACE_LINK_LIBRARIES
-                              $<$<CONFIG:{{config}}>:{{lib_info["requires"]}}>)
+        target_link_libraries({{lib}} INTERFACE {{lib_info["requires"]}})
         {% endif %}
 
         {% endfor %}
diff --git a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
index 3d96b8a053d..07a20126091 100644
--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
@@ -190,8 +190,12 @@ def test_libs(self, matrix_client):
         assert "Conan: Target declared imported STATIC library 'matrix::matrix'" in c.out
 
     def test_libs_components(self, matrix_client_components):
+        """
+        explicit usage of components
+        """
         c = matrix_client_components
 
+        # TODO: Check that find_package(.. COMPONENTS nonexisting) fails
         c.run("new cmake_exe -d name=app -d version=0.1 -d requires=matrix/1.0")
         cmake = textwrap.dedent("""
             cmake_minimum_required(VERSION 3.15)
@@ -210,3 +214,29 @@ def test_libs_components(self, matrix_client_components):
         c.run("build . -c tools.cmake.cmakedeps:new=True")
         assert "Conan: Target declared imported STATIC library 'matrix::vector'" in c.out
         assert "Conan: Target declared imported STATIC library 'matrix::module'" in c.out
+
+    def test_libs_components_default(self, matrix_client_components):
+        """
+        Test that the default components are used when no component is specified
+        """
+        c = matrix_client_components
+
+        c.run("new cmake_exe -d name=app -d version=0.1 -d requires=matrix/1.0")
+
+        app_cpp = textwrap.dedent("""
+            #include "module.h"
+            int main() { module();}
+            """)
+        cmake = textwrap.dedent("""
+            cmake_minimum_required(VERSION 3.15)
+            project(app CXX)
+
+            find_package(matrix CONFIG REQUIRED)
+            add_executable(app src/app.cpp)
+            target_link_libraries(app PRIVATE matrix::matrix)
+            """)
+        c.save({"src/app.cpp": app_cpp,
+                "CMakeLists.txt": cmake})
+        c.run("build . -c tools.cmake.cmakedeps:new=True")
+        assert "Conan: Target declared imported STATIC library 'matrix::vector'" in c.out
+        assert "Conan: Target declared imported STATIC library 'matrix::module'" in c.out
