diff --git a/conan/tools/cmake/cmakedeps2/config.py b/conan/tools/cmake/cmakedeps2/config.py
index 494868ed3c5..a59f6319e9d 100644
--- a/conan/tools/cmake/cmakedeps2/config.py
+++ b/conan/tools/cmake/cmakedeps2/config.py
@@ -44,18 +44,10 @@ def _template(self):
         endif()
 
         include(${CMAKE_CURRENT_LIST_DIR}/{{ targets_include_file }})
-        include(CMakeFindDependencyMacro)
 
         get_property(isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
         if(NOT isMultiConfig AND NOT CMAKE_BUILD_TYPE)
            message(FATAL_ERROR "Please, set the CMAKE_BUILD_TYPE variable when calling to CMake "
                                "adding the '-DCMAKE_BUILD_TYPE=<build_type>' argument.")
         endif()
-
-        foreach(_DEPENDENCY {{ pkg_var(pkg_name, 'FIND_DEPENDENCY_NAMES', '') }} )
-            # Check that we have not already called a find_package with the transitive dependency
-            if(NOT {{ '${_DEPENDENCY}' }}_FOUND)
-                find_dependency({{ '${_DEPENDENCY}' }} REQUIRED ${${_DEPENDENCY}_FIND_MODE})
-            endif()
-        endforeach()
         """)
diff --git a/conan/tools/cmake/cmakedeps2/target_configuration.py b/conan/tools/cmake/cmakedeps2/target_configuration.py
index 770cf38b6ce..bc7e3a3f55e 100644
--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -19,7 +19,6 @@ def __init__(self, cmakedeps, conanfile):
     def content(self):
         t = Template(self._template, trim_blocks=True, lstrip_blocks=True,
                      undefined=jinja2.StrictUndefined)
-        print(t.render(self._context))
         return t.render(self._context)
 
     @property
@@ -33,7 +32,6 @@ def _requires(self, info, components):
         pkg_name = self._conanfile.ref.name
         transitive_reqs = self._cmakedeps.get_transitive_requires(self._conanfile)
         if not requires and not components:  # global cpp_info without components definition
-            # TODO: Not tested at all
             for dependency in transitive_reqs.values():
                 dep_pkg_name = dependency.ref.name
                 dep_cpp_info = dependency.cpp_info
@@ -41,11 +39,11 @@ def _requires(self, info, components):
                     for name, component in dep_cpp_info.components.items():
                         result.append(f"{dep_pkg_name}::{name}")
                 else:
-                    result.append("{dep_pkg_name}::{dep_pkg_name}")
+                    result.append(f"{dep_pkg_name}::{dep_pkg_name}")
             return result
 
-        for required_scope, required_comp in requires:
-            if required_scope is None:  # Points to a component of same package
+        for required_pkg, required_comp in requires:
+            if required_pkg is None:  # Points to a component of same package
                 dep_component = components[required_comp]
                 dep_target = dep_component.get_property("cmake_target_name") or f"{pkg_name}::{required_comp}"
                 result.append(dep_target)
@@ -53,15 +51,17 @@ def _requires(self, info, components):
                 try:  # Make sure the declared dependency is at least in the recipe requires
                     self._conanfile.dependencies[required_pkg]
                 except KeyError:
-                    raise ConanException(f"{self.conanfile}: component '{comp_name}' required "
+                    raise ConanException(f"{self._conanfile}: component '{required_comp}' required "
                                          f"'{required_pkg}::{required_comp}', "
                                          f"but '{required_pkg}' is not a direct dependency")
                 try:
-                    req = transitive_requires[required_pkg]
+                    req = transitive_reqs[required_pkg]
                 except KeyError:  # The transitive dep might have been skipped
                     pass
                 else:
-                    public_comp_deps.append(self.get_component_alias(req, required_comp))
+                    # TODO: Missing cmake_target_name for req
+                    dep_target = f"{required_pkg}::{required_comp}"
+                    result.append(dep_target)
         return result
 
     @property
@@ -121,12 +121,26 @@ def _add_exe(info):
             if cpp_info.exe:
                 _add_exe(cpp_info)
 
-        return {"package_folder": package_folder,
+        # TODO: Missing find_modes
+        dependencies = self._get_dependencies()
+        return {"dependencies": dependencies,
+                "package_folder": package_folder,
                 "package_folder_var": package_folder_var,
                 "config": config,
                 "exes": exes,
                 "libs": libs}
 
+    def _get_dependencies(self):
+        """ transitive dependencies Filenames for find_dependency()
+        """
+        # TODO: Filter build requires
+        # if self._require.build:
+        #    return []
+
+        transitive_reqs = self._cmakedeps.get_transitive_requires(self._conanfile)
+        ret = {self._cmakedeps.get_cmake_filename(r): None for r in transitive_reqs.values()}
+        return ret
+
     @staticmethod
     def _path(p, package_folder, package_folder_var):
         def escape(p_):
@@ -146,8 +160,18 @@ def _template(self):
         return textwrap.dedent("""\
         set({{package_folder_var}} "{{package_folder}}")
 
-        # Libs information
+        # Dependencies finding
+        include(CMakeFindDependencyMacro)
+
+        {% for dep, dep_find_mode in dependencies.items() %}
+        if(NOT {{dep}}_FOUND)
+            find_dependency({{dep}} REQUIRED {{dep_find_mode}})
+        endif()
+        {% endfor %}
+
+        ################# Libs information ##############
         {% for lib, lib_info in libs.items() %}
+        #################### {{lib}} ####################
         message(STATUS "Conan: Target declared imported {{lib_info["type"]}} library '{{lib}}'")
         if(NOT TARGET {{ lib }})
             add_library({{lib}} {{lib_info["type"]}} IMPORTED)
@@ -171,8 +195,9 @@ def _template(self):
 
         {% endfor %}
 
-        # Exe information
+        ################# Exes information ##############
         {% for exe, location in exes.items() %}
+        #################### {{exe}} ####################
         message(STATUS "Conan: Target declared imported executable '{{exe}}'")
         if(NOT TARGET {{ exe }})
             add_executable({{exe}} IMPORTED)
diff --git a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
index 07a20126091..e66c1e1d7f8 100644
--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
@@ -189,6 +189,12 @@ def test_libs(self, matrix_client):
         c.run("build . -c tools.cmake.cmakedeps:new=True")
         assert "Conan: Target declared imported STATIC library 'matrix::matrix'" in c.out
 
+    def test_libs_transitive(self, transitive_libraries):
+        c = transitive_libraries
+        c.run("new cmake_lib -d name=app -d version=0.1 -d requires=engine/1.0")
+        c.run("build . -c tools.cmake.cmakedeps:new=True")
+        assert "Conan: Target declared imported STATIC library 'matrix::matrix'" in c.out
+
     def test_libs_components(self, matrix_client_components):
         """
         explicit usage of components
@@ -240,3 +246,129 @@ def test_libs_components_default(self, matrix_client_components):
         c.run("build . -c tools.cmake.cmakedeps:new=True")
         assert "Conan: Target declared imported STATIC library 'matrix::vector'" in c.out
         assert "Conan: Target declared imported STATIC library 'matrix::module'" in c.out
+
+    def test_libs_components_transitive(self, matrix_client_components):
+        """
+        explicit usage of components
+        matrix::module -> matrix::vector
+
+        engine::bots -> engine::physix
+        engine::physix -> matrix::vector
+        engine::world -> engine::physix, matrix::module
+        """
+        c = matrix_client_components
+
+        from conan.test.assets.sources import gen_function_h
+        bots_h = gen_function_h(name="bots")
+        from conan.test.assets.sources import gen_function_cpp
+        bots_cpp = gen_function_cpp(name="bots", includes=["bots", "physix"], calls=["physix"])
+        physix_h = gen_function_h(name="physix")
+        physix_cpp = gen_function_cpp(name="physix", includes=["physix", "vector"], calls=["vector"])
+        world_h = gen_function_h(name="world")
+        world_cpp = gen_function_cpp(name="world", includes=["world", "physix", "module"],
+                                     calls=["physix", "module"])
+
+        conanfile = textwrap.dedent("""
+            from os.path import join
+            from conan import ConanFile
+            from conan.tools.cmake import CMake
+            from conan.tools.files import copy
+
+            class Engine(ConanFile):
+              name = "engine"
+              version = "1.0"
+              settings = "os", "compiler", "build_type", "arch"
+              generators = "CMakeToolchain"
+              exports_sources = "src/*", "CMakeLists.txt"
+
+              requires = "matrix/1.0"
+              generators = "CMakeDeps", "CMakeToolchain"
+
+              def build(self):
+                  cmake = CMake(self)
+                  cmake.configure()
+                  cmake.build()
+
+              def package(self):
+                  cmake = CMake(self)
+                  cmake.install()
+
+              def package_info(self):
+                  self.cpp_info.components["bots"].libs = ["bots"]
+                  self.cpp_info.components["bots"].includedirs = ["include"]
+                  self.cpp_info.components["bots"].libdirs = ["lib"]
+                  self.cpp_info.components["bots"].requires = ["physix"]
+
+                  self.cpp_info.components["physix"].libs = ["physix"]
+                  self.cpp_info.components["physix"].includedirs = ["include"]
+                  self.cpp_info.components["physix"].libdirs = ["lib"]
+                  self.cpp_info.components["physix"].requires = ["matrix::vector"]
+
+                  self.cpp_info.components["world"].libs = ["world"]
+                  self.cpp_info.components["world"].includedirs = ["include"]
+                  self.cpp_info.components["world"].libdirs = ["lib"]
+                  self.cpp_info.components["world"].requires = ["physix", "matrix::module"]
+                  """)
+
+        cmakelists = textwrap.dedent("""
+               set(CMAKE_CXX_COMPILER_WORKS 1)
+               set(CMAKE_CXX_ABI_COMPILED 1)
+               cmake_minimum_required(VERSION 3.15)
+               project(matrix CXX)
+
+               find_package(matrix CONFIG REQUIRED)
+
+               add_library(physix src/physix.cpp)
+               add_library(bots src/bots.cpp)
+               add_library(world src/world.cpp)
+
+               target_link_libraries(physix PRIVATE matrix::vector)
+               target_link_libraries(bots PRIVATE physix)
+               target_link_libraries(world PRIVATE physix matrix::module)
+
+               set_target_properties(bots PROPERTIES PUBLIC_HEADER "src/bots.h")
+               set_target_properties(physix PROPERTIES PUBLIC_HEADER "src/physix.h")
+               set_target_properties(world PROPERTIES PUBLIC_HEADER "src/world.h")
+               install(TARGETS physix bots world)
+               """)
+        c.save({"src/physix.h": physix_h,
+                "src/physix.cpp": physix_cpp,
+                "src/bots.h": bots_h,
+                "src/bots.cpp": bots_cpp,
+                "src/world.h": world_h,
+                "src/world.cpp": world_cpp,
+                "CMakeLists.txt": cmakelists,
+                "conanfile.py": conanfile})
+        c.run("create .")
+
+        c.save({}, clean_first=True)
+        c.run("new cmake_exe -d name=app -d version=0.1 -d requires=engine/1.0")
+        cmake = textwrap.dedent("""
+            cmake_minimum_required(VERSION 3.15)
+            project(app CXX)
+
+            find_package(engine CONFIG REQUIRED)
+
+            add_executable(app src/app.cpp)
+            target_link_libraries(app PRIVATE engine::bots)
+
+            install(TARGETS app)
+            """)
+        app_cpp = textwrap.dedent("""
+            #include "bots.h"
+            int main() { bots();}
+            """)
+        c.save({"CMakelists.txt": cmake,
+                "src/app.cpp": app_cpp})
+        c.run("create . -c tools.cmake.cmakedeps:new=True")
+        assert "Conan: Target declared imported STATIC library 'matrix::vector'" in c.out
+        assert "Conan: Target declared imported STATIC library 'matrix::module'" in c.out
+        assert "Conan: Target declared imported INTERFACE library 'matrix::matrix'" in c.out
+        assert "Conan: Target declared imported STATIC library 'engine::bots'" in c.out
+        assert "Conan: Target declared imported STATIC library 'engine::physix'" in c.out
+        assert "Conan: Target declared imported STATIC library 'engine::world'" in c.out
+        assert "Conan: Target declared imported INTERFACE library 'engine::engine'" in c.out
+
+        assert "bots: Release!" in c.out
+        assert "physix: Release!" in c.out
+        assert "vector: Release!" in c.out
diff --git a/test/functional/toolchains/conftest.py b/test/functional/toolchains/conftest.py
index 195e6d00e53..d00a5a51175 100644
--- a/test/functional/toolchains/conftest.py
+++ b/test/functional/toolchains/conftest.py
@@ -106,6 +106,7 @@ def package_info(self):
        target_link_libraries(module PRIVATE vector)
 
        set_target_properties(module PROPERTIES PUBLIC_HEADER "src/module.h")
+       set_target_properties(vector PROPERTIES PUBLIC_HEADER "src/vector.h")
        install(TARGETS vector module)
        """)
     c.save({"src/vector.h": vector_h,
