diff --git a/conan/tools/cmake/cmakedeps2/target_configuration.py b/conan/tools/cmake/cmakedeps2/target_configuration.py
index 4755bf1a9b7..5832efa65dc 100644
--- a/conan/tools/cmake/cmakedeps2/target_configuration.py
+++ b/conan/tools/cmake/cmakedeps2/target_configuration.py
@@ -75,8 +75,15 @@ def _context(self):
         libs = {}
 
         def _add_libs(info, component_name=None):
+            defines = " ".join(info.defines)
+            includedirs = ";".join(self._path(i, package_folder, package_folder_var)
+                                   for i in info.includedirs) if info.includedirs else ""
+            requires = " ".join(self._requires(info, self._conanfile.cpp_info.components))
+            cxxflags = " ".join(info.cxxflags)
             if info.libs:  # TODO: Handle component_name for libs
-                assert len(info.libs) == 1, "New CMakeDeps only allows 1 lib per component"
+                if len(info.libs) != 1:
+                    raise ConanException(f"New CMakeDeps only allows 1 lib per component:\n"
+                                         f"{self._conanfile}: {info.libs}")
                 lib_name = info.libs[0]
                 target_name = info.get_property("cmake_target_name") or f"{pkg_name}::{lib_name}"
                 info.deduce_cps(self._conanfile.package_type)
@@ -84,22 +91,23 @@ def _add_libs(info, component_name=None):
                 lib_type = "SHARED" if info.type is PackageType.SHARED else \
                     "STATIC" if info.type is PackageType.STATIC else \
                     "INTERFACE" if info.type is PackageType.HEADER else None
-                includedirs = ";".join(self._path(i, package_folder, package_folder_var)
-                                       for i in info.includedirs) if info.includedirs else ""
-                requires = " ".join(self._requires(info, self._conanfile.cpp_info.components))
+
                 if lib_type:
                     libs[target_name] = {"type": lib_type,
                                          "includedirs": includedirs,
                                          "location": location,
                                          "link_location": "",
-                                         "requires": requires}
+                                         "requires": requires,
+                                         "defines": defines,
+                                         "cxxflags": cxxflags}
             elif info.includedirs and not info.exe:  # Pure header only
                 cmp_name = component_name or pkg_name
                 target_name = info.get_property("cmake_target_name") or f"{pkg_name}::{cmp_name}"
-                includedirs = ";".join(self._path(i, package_folder, package_folder_var)
-                                       for i in info.includedirs)
                 libs[target_name] = {"type": "INTERFACE",
-                                     "includedirs": includedirs}
+                                     "includedirs": includedirs,
+                                     "defines": defines,
+                                     "requires": requires,
+                                     "cxxflags": cxxflags}
 
         if cpp_info.has_components:
             for name, component in cpp_info.components.items():
@@ -169,6 +177,11 @@ def escape(p_):
             return escape(p)
         return f"${{{package_folder_var}}}/{escape(p)}"
 
+    @staticmethod
+    def _escape_cmake_string(values):
+        return " ".join(v.replace("\\", "\\\\").replace('$', '\\$').replace('"', '\\"')
+                        for v in values)
+
     @property
     def _template(self):
         # TODO: Check why not set_property instead of target_link_libraries
@@ -195,6 +208,15 @@ def _template(self):
         set_property(TARGET {{lib}} APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES
                      $<$<CONFIG:{{config}}>:{{lib_info["includedirs"]}}>)
         {% endif %}
+        {% if lib_info.get("defines") %}
+        set_property(TARGET {{lib}} APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS
+                     $<$<CONFIG:{{config}}>:{{lib_info["defines"]}}>)
+        {% endif %}
+        {% if lib_info.get("cxxflags") %}
+        set_property(TARGET {{lib}} APPEND PROPERTY INTERFACE_COMPILE_OPTIONS
+                     $<$<COMPILE_LANGUAGE:CXX>:$<$<CONFIG:{{config}}>:{{lib_info["cxxflags"]}}>>)
+        {% endif %}
+
         {% if lib_info["type"] != "INTERFACE" %}
         set_property(TARGET {{lib}} APPEND PROPERTY IMPORTED_CONFIGURATIONS {{config}})
         set_target_properties({{lib}} PROPERTIES IMPORTED_LOCATION_{{config}}
diff --git a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
index 0ef0366a7fa..793989c0c74 100644
--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new.py
@@ -1,3 +1,4 @@
+import platform
 import textwrap
 
 import pytest
@@ -222,6 +223,9 @@ def test_libs_components(self, matrix_client_components):
         c.run("build . -c tools.cmake.cmakedeps:new=True")
         assert "Conan: Target declared imported STATIC library 'matrix::vector'" in c.out
         assert "Conan: Target declared imported STATIC library 'matrix::module'" in c.out
+        if platform.system() == "Windows":
+            c.run_command(r".\build\Release\app.exe")
+            assert "Matrix headers __cplusplus: __cplusplus2014" in c.out
 
     def test_libs_components_default(self, matrix_client_components):
         """
diff --git a/test/functional/toolchains/conftest.py b/test/functional/toolchains/conftest.py
index fbb73dc4350..31b60f8b807 100644
--- a/test/functional/toolchains/conftest.py
+++ b/test/functional/toolchains/conftest.py
@@ -60,7 +60,14 @@ def _matrix_client_components():
     c = TestClient()
     headers_h = textwrap.dedent("""
         #include <iostream>
-        void headers(){ std::cout << "Matrix headers: Release!" << std::endl; }
+        #ifndef MY_MATRIX_HEADERS_DEFINE
+        #error "Fatal error MY_MATRIX_HEADERS_DEFINE not defined"
+        #endif
+        void headers(){ std::cout << "Matrix headers: Release!" << std::endl;
+            #if __cplusplus
+            std::cout << "  Matrix headers __cplusplus: __cplusplus" << __cplusplus << std::endl;
+            #endif
+        }
         """)
     vector_h = gen_function_h(name="vector")
     vector_cpp = gen_function_cpp(name="vector", includes=["vector"])
@@ -92,6 +99,9 @@ def package_info(self):
 
                 self.cpp_info.components["headers"].includedirs = ["include/headers"]
                 self.cpp_info.components["headers"].set_property("cmake_target_name", "MatrixHeaders")
+                self.cpp_info.components["headers"].defines = ["MY_MATRIX_HEADERS_DEFINE=1"]
+                if self.settings.compiler == "msvc":
+                    self.cpp_info.components["headers"].cxxflags = ["/Zc:__cplusplus"]
 
                 self.cpp_info.components["vector"].libs = ["vector"]
                 self.cpp_info.components["vector"].includedirs = ["include"]
