diff --git a/conans/model/requires.py b/conans/model/requires.py
index d9a9dbc7646..cb819d74e46 100644
--- a/conans/model/requires.py
+++ b/conans/model/requires.py
@@ -254,6 +254,7 @@ def aggregate(self, other):
         self.transitive_libs = self.transitive_libs or other.transitive_libs
         if not other.test:
             self.test = False  # it it was previously a test, but also required by non-test
+        self.is_test = self.is_test or other.is_test
         # package_id_mode is not being propagated downstream. So it is enough to check if the
         # current require already defined it or not
         if self.package_id_mode is None:
@@ -344,6 +345,7 @@ def transform_downstream(self, pkg_type, require, dep_pkg_type):
 
         if self.test:
             downstream_require.test = True
+        downstream_require.is_test = require.is_test
 
         # If the current one is resolving conflicts, the downstream one will be too
         downstream_require.force = require.force
diff --git a/test/integration/graph/test_test_requires.py b/test/integration/graph/test_test_requires.py
index f85391d9df5..5342bc3ad3b 100644
--- a/test/integration/graph/test_test_requires.py
+++ b/test/integration/graph/test_test_requires.py
@@ -198,6 +198,38 @@ def package_info(self):
     assert "libc/0.1: Created package" in c.out
 
 
+def test_requires_transitive_diamond_components_order():
+    """
+    like the above, but in different order
+     libc  --(test-requires)---> liba
+       |-----> libb -----------/
+    https://github.com/conan-io/conan/issues/17164
+    """
+    c = TestClient(light=True)
+    libc = textwrap.dedent("""
+        from conan import ConanFile
+        class LibC(ConanFile):
+            name = "libc"
+            version = "0.1"
+
+            def requirements(self):
+                self.test_requires("liba/1.0")
+                self.requires("libb/1.0")
+
+            def package_info(self):
+                 self.cpp_info.components["comp"].libs = ["libc"]
+                 self.cpp_info.components["comp"].requires.append("libb::libb")
+        """)
+    c.save({"liba/conanfile.py": GenConanfile("liba", "1.0"),
+            "libb/conanfile.py": GenConanfile("libb", "1.0").with_requires("liba/1.0"),
+            "libc/conanfile.py": libc})
+    c.run("create liba")
+    c.run("create libb")
+    c.run("create libc")
+    # This used to crash due to component not defined to liba
+    assert "libc/0.1: Created package" in c.out
+
+
 def test_test_requires_options():
     """ the default_options = {} values also propagate to ``test_requires()`` """
     c = TestClient(light=True)
