diff --git a/kubernetes/base/config/exec_provider.py b/kubernetes/base/config/exec_provider.py
index dad60f9543..03dc0386ef 100644
--- a/kubernetes/base/config/exec_provider.py
+++ b/kubernetes/base/config/exec_provider.py
@@ -54,7 +54,7 @@ def __init__(self, exec_config, cwd, cluster=None):
                 additional_vars[name] = value
             self.env.update(additional_vars)
         if exec_config.safe_get('provideClusterInfo'):
-            self.cluster = cluster.value
+            self.cluster = cluster
         else:
             self.cluster = None
         self.cwd = cwd or None
@@ -71,7 +71,7 @@ def run(self, previous_response=None):
         if previous_response:
             kubernetes_exec_info['spec']['response'] = previous_response
         if self.cluster:
-            kubernetes_exec_info['spec']['cluster'] = self.cluster
+            kubernetes_exec_info['spec']['cluster'] = self.cluster.value
 
         self.env['KUBERNETES_EXEC_INFO'] = json.dumps(kubernetes_exec_info)
         process = subprocess.Popen(
