diff --git a/changelog.d/1610.bugfix.md b/changelog.d/1610.bugfix.md
new file mode 100644
index 000000000..4ef0daa72
--- /dev/null
+++ b/changelog.d/1610.bugfix.md
@@ -0,0 +1 @@
+Fix incorrect order of flags when using `pipx upgrade`.
\ No newline at end of file
diff --git a/src/pipx/venv.py b/src/pipx/venv.py
index 422de3da0..4f64a4c25 100644
--- a/src/pipx/venv.py
+++ b/src/pipx/venv.py
@@ -441,7 +441,7 @@ def has_package(self, package_name: str) -> bool:
     def upgrade_package_no_metadata(self, package_name: str, pip_args: List[str]) -> None:
         logger.info("Upgrading %s", package_descr := full_package_description(package_name, package_name))
         with animate(f"upgrading {package_descr}", self.do_animation):
-            pip_process = self._run_pip(["--no-input", "install"] + pip_args + ["--upgrade", package_name])
+            pip_process = self._run_pip(["--no-input", "install", "--upgrade"] + pip_args + [package_name])
         subprocess_post_check(pip_process)
 
     def upgrade_package(
@@ -456,7 +456,7 @@ def upgrade_package(
     ) -> None:
         logger.info("Upgrading %s", package_descr := full_package_description(package_name, package_or_url))
         with animate(f"upgrading {package_descr}", self.do_animation):
-            pip_process = self._run_pip(["--no-input", "install"] + pip_args + ["--upgrade", package_or_url])
+            pip_process = self._run_pip(["--no-input", "install", "--upgrade"] + pip_args + [package_or_url])
         subprocess_post_check(pip_process)
 
         self.update_package_metadata(
