diff --git a/tests/isolated/fork_in_thread.py b/tests/isolated/fork_in_thread.py
index 29fff56a2..68aa0e75f 100644
--- a/tests/isolated/fork_in_thread.py
+++ b/tests/isolated/fork_in_thread.py
@@ -29,6 +29,7 @@ def forker():
     else:
         global parent
         parent = False
+    results.append(True)
 
 t = threading.Thread(target=background)
 t.start()
@@ -38,6 +39,6 @@ def forker():
 t.join()
 
 check_current()
-assert results == [True]
+assert results == [True, True]
 if parent:
     print("pass")
