diff --git a/test/test_InfoExtractor.py b/test/test_InfoExtractor.py
index 07295d058726..d275b067c4e1 100644
--- a/test/test_InfoExtractor.py
+++ b/test/test_InfoExtractor.py
@@ -638,6 +638,7 @@ def test_parse_m3u8_formats(self):
                 'img_bipbop_adv_example_fmp4',
                 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/master.m3u8',
                 [{
+                    # 60kbps (bitrate not provided in m3u8); sorted as worst because it's grouped with lowest bitrate video track
                     'format_id': 'aud1-English',
                     'url': 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/a1/prog_index.m3u8',
                     'manifest_url': 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/master.m3u8',
@@ -646,6 +647,7 @@ def test_parse_m3u8_formats(self):
                     'protocol': 'm3u8_native',
                     'audio_ext': 'mp4',
                 }, {
+                    # 192kbps (bitrate not provided in m3u8)
                     'format_id': 'aud3-English',
                     'url': 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/a3/prog_index.m3u8',
                     'manifest_url': 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/master.m3u8',
@@ -654,7 +656,7 @@ def test_parse_m3u8_formats(self):
                     'protocol': 'm3u8_native',
                     'audio_ext': 'mp4',
                 }, {
-                    # Sorted as best audio because it is grouped with the highest bitrate video track
+                    # 384kbps (bitrate not provided in m3u8); sorted as best because it's grouped with the highest bitrate video track
                     'format_id': 'aud2-English',
                     'url': 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/a2/prog_index.m3u8',
                     'manifest_url': 'https://devstreaming-cdn.apple.com/videos/streaming/examples/img_bipbop_adv_example_fmp4/master.m3u8',
