diff --git a/beetsplug/lastgenre/__init__.py b/beetsplug/lastgenre/__init__.py
index f0c13c4964..1a56d37798 100644
--- a/beetsplug/lastgenre/__init__.py
+++ b/beetsplug/lastgenre/__init__.py
@@ -254,6 +254,8 @@ def fetch_genre(self, lastfm_obj):
 
     def _filter_valid_genres(self, genres: list[str]) -> list[str]:
         """Filter list of genres, only keep valid."""
+        if not genres:
+            return []
         return [x for x in genres if self._is_valid(x)]
 
     def _is_valid(self, genre: str) -> bool:
