diff --git a/beetsplug/lastgenre/__init__.py b/beetsplug/lastgenre/__init__.py
index 1a56d37798..30b44e1875 100644
--- a/beetsplug/lastgenre/__init__.py
+++ b/beetsplug/lastgenre/__init__.py
@@ -441,11 +441,12 @@ def _get_genre(
                     label = f"keep + {label}"
                 return self._format_and_stringify(resolved_genres), label
 
-        # Nothing found, leave original.
-        if obj.genre:
-            return obj.genre, "original fallback"
+        # Nothing found, leave original if configured and valid.
+        if obj.genre and self.config["keep_existing"]:
+            if not self.whitelist or self._is_valid(obj.genre.lower()):
+                return obj.genre, "original fallback"
 
-        # No original, return fallback string.
+        # Return fallback string.
         if fallback := self.config["fallback"].get():
             return fallback, "fallback"
 
diff --git a/test/plugins/test_lastgenre.py b/test/plugins/test_lastgenre.py
index ad2dedf5ba..49d219de98 100644
--- a/test/plugins/test_lastgenre.py
+++ b/test/plugins/test_lastgenre.py
@@ -292,7 +292,8 @@ def test_sort_by_depth(self):
             },
             ("Unknown Genre, Jazz", "keep + artist, any"),
         ),
-        # 7 - fallback to any original when nothing found and whitelist disabled
+        # 7 - Keep the original genre when force and keep_existing are on, and
+        # whitelist is disabled
         (
             {
                 "force": True,
@@ -303,6 +304,46 @@ def test_sort_by_depth(self):
                 "canonical": False,
                 "prefer_specific": False,
             },
+            "any existing",
+            {
+                "track": None,
+                "album": None,
+                "artist": None,
+            },
+            ("any existing", "original fallback"),
+        ),
+        # 7.1 - Keep the original genre when force and keep_existing are on, and
+        # whitelist is enabled, and genre is valid.
+        (
+            {
+                "force": True,
+                "keep_existing": True,
+                "source": "track",
+                "whitelist": True,
+                "fallback": "fallback genre",
+                "canonical": False,
+                "prefer_specific": False,
+            },
+            "Jazz",
+            {
+                "track": None,
+                "album": None,
+                "artist": None,
+            },
+            ("Jazz", "original fallback"),
+        ),
+        # 7.2 - Return the configured fallback when force is on but
+        # keep_existing is not.
+        (
+            {
+                "force": True,
+                "keep_existing": False,
+                "source": "track",
+                "whitelist": True,
+                "fallback": "fallback genre",
+                "canonical": False,
+                "prefer_specific": False,
+            },
             "Jazz",
             {
                 "track": None,
