diff --git a/src/tox/config/cli/parser.py b/src/tox/config/cli/parser.py
index 7afe87753e..24e06b5a79 100644
--- a/src/tox/config/cli/parser.py
+++ b/src/tox/config/cli/parser.py
@@ -232,6 +232,7 @@ def __call__(
             dest="discover",
             nargs="+",
             metavar="path",
+            of_type=list[str],
             help="for Python discovery first try the Python executables under these paths",
             default=[],
         )
diff --git a/tests/config/cli/test_cli_env_var.py b/tests/config/cli/test_cli_env_var.py
index 3c870ae9e7..7f265c78ce 100644
--- a/tests/config/cli/test_cli_env_var.py
+++ b/tests/config/cli/test_cli_env_var.py
@@ -84,6 +84,7 @@ def test_env_var_exhaustive_parallel_values(
     monkeypatch.setenv("TOX_PARALLEL", "3")
     monkeypatch.setenv("TOX_PARALLEL_LIVE", "no")
     monkeypatch.setenv("TOX_OVERRIDE", "a=b;c=d")
+    monkeypatch.setenv("TOX_DISCOVER", "/foo/bar;/bar/baz;/baz/foo")
 
     options = get_options()
     assert vars(options.parsed) == {
@@ -93,7 +94,7 @@ def test_env_var_exhaustive_parallel_values(
         "default_runner": "virtualenv",
         "develop": False,
         "devenv_path": None,
-        "discover": [],
+        "discover": ["/foo/bar", "/bar/baz", "/baz/foo"],
         "env": CliEnv(["py37", "py36"]),
         "force_dep": [],
         "hash_seed": ANY,
