diff --git a/pvlib/tests/test_modelchain.py b/pvlib/tests/test_modelchain.py
index 5678470f10..72c349c8e6 100644
--- a/pvlib/tests/test_modelchain.py
+++ b/pvlib/tests/test_modelchain.py
@@ -1320,6 +1320,7 @@ def test_temperature_model_not_specified():
         _ = ModelChain(system, location,
                        aoi_model='no_loss', spectral_model='no_loss')
 
+
 def test_dc_model_user_func(pvwatts_dc_pvwatts_ac_system, location, weather,
                             mocker):
     m = mocker.spy(sys.modules[__name__], 'poadc')
