diff --git a/sphinx/util/inspect.py b/sphinx/util/inspect.py
index 5790a18a6b3..0e31a50d28a 100644
--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -274,7 +274,11 @@ def is_builtin_classmethod(obj: Any, cls: Any = None, name: str | None = None) -
         for basecls in getmro(cls):
             meth = basecls.__dict__.get(name, sentinel)
             if meth is not sentinel:
-                return is_builtin_classmethod(meth)
+                return is_classmethod_descriptor(obj, cls, name) or (
+                    isbuiltin(obj)
+                    and getattr(obj, '__self__', None) is not None
+                    and isclass(obj.__self__)
+                )
     return False
 
 
