diff --git a/sphinx/util/inspect.py b/sphinx/util/inspect.py
index ac22ce9bacc..4818e32794b 100644
--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -259,7 +259,9 @@ def is_classmethod_descriptor(
     return False
 
 
-def is_builtin_classmethod_like(obj: Any, cls: Any = None, name: str | None = None) -> bool:
+def is_builtin_classmethod_like(
+    obj: Any, cls: Any = None, name: str | None = None
+) -> bool:
     """Check if the object looks like a class method implemented in C.
 
     This is equivalent to test that *obj* is a class method descriptor
diff --git a/tests/test_util/test_util_inspect.py b/tests/test_util/test_util_inspect.py
index bc1d89a43b5..11509006b6d 100644
--- a/tests/test_util/test_util_inspect.py
+++ b/tests/test_util/test_util_inspect.py
@@ -67,7 +67,7 @@ def from_bytes(cls, *a, **kw):
 
     def conjugate(self):
         return super().conjugate()
-    
+
 
 def func():
     pass
@@ -724,7 +724,7 @@ class Qux:
         (False, Inherited, 'meth'),
         (False, MyInt, 'conjugate'),
         (False, MyIntOverride, 'conjugate'),
-    ]
+    ],
 )
 def test_isclassmethod(expect, klass, name):
     subject = getattr(klass, name)
@@ -746,7 +746,7 @@ def test_isclassmethod(expect, klass, name):
         (False, int, 'conjugate'),
         (False, MyInt, 'classmeth'),
         (False, MyIntOverride, 'from_bytes'),  # overridden in pure Python
-    ]
+    ],
 )
 def test_is_classmethod_descriptor(expect, klass, dict_key):
     in_dict = dict_key in klass.__dict__
@@ -772,7 +772,7 @@ def test_is_classmethod_descriptor(expect, klass, dict_key):
         (False, int, 'conjugate'),
         (False, MyInt, 'classmeth'),
         (False, MyIntOverride, 'from_bytes'),  # overridden in pure Python
-    ]
+    ],
 )
 def test_is_builtin_classmethod_like(expect, klass, dict_key):
     method = getattr(klass, dict_key)
@@ -797,7 +797,7 @@ def test_is_builtin_classmethod_like(expect, klass, dict_key):
         (False, int, '__init__'),
         (False, int, 'conjugate'),
         (False, MyIntOverride, 'conjugate'),  # overridden in pure Python
-    ]
+    ],
 )
 def test_is_classmethod_like(expect, klass, name):
     subject = getattr(klass, name)
