diff --git a/test/unit/rules/resources/stepfunctions/test_state_machine_definition.py b/test/unit/rules/resources/stepfunctions/test_state_machine_definition.py
index 00fdcf4ba6..74957de20b 100644
--- a/test/unit/rules/resources/stepfunctions/test_state_machine_definition.py
+++ b/test/unit/rules/resources/stepfunctions/test_state_machine_definition.py
@@ -95,7 +95,7 @@ def rule():
                         "End": True,
                         "ItemsPath": "$",
                         "Parameters": {"BatchNumber.$": "$$.Map.Item.Value"},
-                        "Iterator": {
+                        "ItemProcessor": {
                             "StartAt": "Submit Batch Job",
                             "States": {
                                 "Submit Batch Job": {
@@ -257,7 +257,7 @@ def rule():
                             "MessageNumber.$": "$$.Map.Item.Index",
                             "MessageDetails.$": "$$.Map.Item.Value",
                         },
-                        "Iterator": {
+                        "ItemProcessor": {
                             "StartAt": "Write message to DynamoDB",
                             "States": {
                                 "Write message to DynamoDB": {
