diff --git a/tests/isolated/patcher_fork_after_monkey_patch.py b/tests/isolated/patcher_fork_after_monkey_patch.py
index 421588518..6db70b004 100644
--- a/tests/isolated/patcher_fork_after_monkey_patch.py
+++ b/tests/isolated/patcher_fork_after_monkey_patch.py
@@ -6,7 +6,6 @@
 def check(n, mod, tag):
     assert len(mod._active) == n, 'Expected {} {} threads, got {}'.format(n, tag, mod._active)
 
-
 if __name__ == '__main__':
     import eventlet
     import eventlet.patcher
@@ -17,8 +16,12 @@ def check(n, mod, tag):
     _threading = eventlet.patcher.original('threading')
     import eventlet.green.threading
 
+    global threads_keep_running
+    threads_keep_running = True
+
     def target():
-        eventlet.sleep(0.1)
+        while threads_keep_running:
+            eventlet.sleep(0.001)
 
     threads = [
         threading.Thread(target=target, name='patched'),
@@ -35,20 +38,27 @@ def target():
     check(3, _threading, 'pre-fork original')
     check(5, eventlet.green.threading, 'pre-fork green')
 
-    if os.fork() == 0:
+    if (pid := os.fork()) == 0:
         # Inside the child, we should only have a main _OS_ thread,
         # but green threads should survive.
         check(5, threading, 'child post-fork patched')
         check(1, _threading, 'child post-fork original')
         check(5, eventlet.green.threading, 'child post-fork green')
+        threads_keep_running = False
         sys.exit()
     else:
-        os.wait()
+        wait_pid, status = os.wait()
+        exit_code = os.waitstatus_to_exitcode(status)
+        assert wait_pid == pid
+        assert exit_code == 0, exit_code
 
+    # We're in the parent now; all threads should survive:
     check(5, threading, 'post-fork patched')
     check(3, _threading, 'post-fork original')
     check(5, eventlet.green.threading, 'post-fork green')
 
+    threads_keep_running = False
+
     for t in threads:
         t.join()
 
