diff --git a/patroni/postgresql/config.py b/patroni/postgresql/config.py
index 13139688f..55d4176f2 100644
--- a/patroni/postgresql/config.py
+++ b/patroni/postgresql/config.py
@@ -617,7 +617,8 @@ def format_dsn(self, params: Dict[str, Any]) -> str:
         def escape(value: Any) -> str:
             return re.sub(r'([\'\\ ])', r'\\\1', str(value))
 
-        key_ver = {'gssencmode': 120000, 'channel_binding': 130000, 'sslnegotiation': 170000}
+        key_ver = {'target_session_attrs': 100000, 'gssencmode': 120000, 'channel_binding': 130000,
+                   'sslpassword': 130000, 'sslcrldir': 140000, 'sslnegotiation': 170000}
         return ' '.join('{0}={1}'.format(kw, escape(params[kw])) for kw in keywords
                         if params.get(kw) is not None and self._postgresql.major_version >= key_ver.get(kw, 0))
 
diff --git a/tests/test_postgresql.py b/tests/test_postgresql.py
index aa46e6751..19ab29680 100644
--- a/tests/test_postgresql.py
+++ b/tests/test_postgresql.py
@@ -1184,14 +1184,24 @@ def test_load_current_server_parameters(self):
                           'gssencmode': 'prefer', 'channel_binding': 'prefer', 'sslnegotiation': 'postgres'}})
 
     def test_format_dsn(self):
-        params = {'host': '1', 'port': 2, 'gssencmode': 'prefer',
-                  'channel_binding': 'prefer', 'sslnegotiation': 'postgres'}
-        self.p._major_version = 100000
+        params = {'host': '1', 'port': 2, 'target_session_attrs': 'read-write', 'gssencmode': 'prefer',
+                  'channel_binding': 'prefer', 'sslpassword': 'pwd', 'sslcrldir': '/', 'sslnegotiation': 'postgres'}
+        self.p._major_version = 90600
         self.assertEqual(self.p.config.format_dsn(params), 'host=1 port=2')
+        self.p._major_version = 100000
+        self.assertEqual(self.p.config.format_dsn(params), 'host=1 port=2 target_session_attrs=read-write')
         self.p._major_version = 120000
-        self.assertEqual(self.p.config.format_dsn(params), 'host=1 port=2 gssencmode=prefer')
+        self.assertEqual(self.p.config.format_dsn(params),
+                         'host=1 port=2 gssencmode=prefer target_session_attrs=read-write')
         self.p._major_version = 130000
-        self.assertEqual(self.p.config.format_dsn(params), 'host=1 port=2 gssencmode=prefer channel_binding=prefer')
+        self.assertEqual(self.p.config.format_dsn(params),
+                         'host=1 port=2 sslpassword=pwd gssencmode=prefer '
+                         'channel_binding=prefer target_session_attrs=read-write')
+        self.p._major_version = 140000
+        self.assertEqual(self.p.config.format_dsn(params),
+                         'host=1 port=2 sslpassword=pwd sslcrldir=/ gssencmode=prefer '
+                         'channel_binding=prefer target_session_attrs=read-write')
         self.p._major_version = 170000
         self.assertEqual(self.p.config.format_dsn(params),
-                         'host=1 port=2 gssencmode=prefer channel_binding=prefer sslnegotiation=postgres')
+                         'host=1 port=2 sslpassword=pwd sslcrldir=/ gssencmode=prefer channel_binding=prefer '
+                         'target_session_attrs=read-write sslnegotiation=postgres')
