diff --git a/src/cfnlint/rules/resources/lmbd/PermissionSourceAccount.py b/src/cfnlint/rules/resources/lmbd/PermissionSourceAccount.py
index f1137c7ba7..043ea94d9e 100644
--- a/src/cfnlint/rules/resources/lmbd/PermissionSourceAccount.py
+++ b/src/cfnlint/rules/resources/lmbd/PermissionSourceAccount.py
@@ -32,16 +32,6 @@ def __init__(self):
             keywords=["Resources/AWS::Lambda::Permission/Properties"],
         )
 
-    def _validate_sub_has_account_id(self, validator: Validator, value: Any) -> bool:
-        value = ensure_list(value)
-
-        if isinstance(value[0], str):
-            if re.search(r":(\d{12}|\${AWS::AccountId}):", value[0]):
-                return True
-
-            return False
-        return True
-
     def _validate_is_gettatt_to_bucket(self, validator: Validator, value: Any) -> bool:
         value = ensure_list(value)[0].split(".")[0]
 
@@ -81,10 +71,7 @@ def validate(
 
             fn_k, fn_v = is_function(source_arn)
             if fn_k is not None:
-                if fn_k == "Fn::Sub":
-                    if self._validate_sub_has_account_id(scenario_validator, fn_v):
-                        continue
-                elif fn_k == "Fn::GetAtt":
+                if fn_k == "Fn::GetAtt":
                     if not self._validate_is_gettatt_to_bucket(
                         scenario_validator, fn_v
                     ):
@@ -96,4 +83,5 @@ def validate(
                 yield ValidationError(
                     "'SourceAccount' is a required property",
                     validator="required",
+                    rule=self,
                 )
diff --git a/test/unit/rules/resources/lmbd/test_permission_source_account.py b/test/unit/rules/resources/lmbd/test_permission_source_account.py
index 3135a86ff1..9d370bcda6 100644
--- a/test/unit/rules/resources/lmbd/test_permission_source_account.py
+++ b/test/unit/rules/resources/lmbd/test_permission_source_account.py
@@ -61,6 +61,7 @@ def template():
                 ValidationError(
                     "'SourceAccount' is a required property",
                     validator="required",
+                    rule=PermissionSourceAccount(),
                 )
             ],
         ),
@@ -85,24 +86,6 @@ def template():
             {
                 "SourceArn": {"Fn::Sub": "arn:${AWS::Partition}:s3:::bucket"},
             },
-            [
-                ValidationError(
-                    "'SourceAccount' is a required property",
-                    validator="required",
-                )
-            ],
-        ),
-        (
-            {
-                "SourceArn": {"Fn::Sub": [[], {}]},
-            },
-            [],
-        ),
-        (
-            {
-                "SourceArn": {"Fn::Sub": "arn:${AWS::Partition}:s3:::bucket"},
-                "SourceAccount": {"Ref": "AWS::AccountId"},
-            },
             [],
         ),
         (
@@ -120,6 +103,7 @@ def template():
                 ValidationError(
                     "'SourceAccount' is a required property",
                     validator="required",
+                    rule=PermissionSourceAccount(),
                 )
             ],
         ),
@@ -182,6 +166,7 @@ def template():
                 ValidationError(
                     "'SourceAccount' is a required property",
                     validator="required",
+                    rule=PermissionSourceAccount(),
                 )
             ],
         ),
