diff --git a/src/tox/config/cli/parser.py b/src/tox/config/cli/parser.py
index 068fbca231..707c7fc8ac 100644
--- a/src/tox/config/cli/parser.py
+++ b/src/tox/config/cli/parser.py
@@ -232,7 +232,7 @@ def __call__(
             dest="discover",
             nargs="+",
             metavar="path",
-            of_type=list[str],
+            of_type=List[str],
             help="for Python discovery first try these Python executables",
             default=[],
         )
